diff --git a/docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md b/docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md
index 6eb3cc0e7b0..d9ddf17ae9a 100644
--- a/docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md	
+++ b/docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md	
@@ -292,6 +292,47 @@ unsafe record struct R(
 }
 ```
 
+## `with()` as a collection expression element is treated as collection construction *arguments*
+
+***Introduced in Visual Studio 2022 version 17.14***
+
+`with(...)` when used as an element in a collection expression is bound as arguments passed to constructor or
+factory method used to create the collection, rather than as an invocation expression of a method named `with`.
+
+To bind to a method named `with`, use `@with` instead.
+
+```cs
+object x, y, z = ...;
+object[] items;
+
+items = [with(x, y), z];  // C#13: call to with() method; C#14: error args not supported for object[]
+items = [@with(x, y), z]; // call to with() method
+
+object with(object a, object b) { ... }
+```
+
+## Collection expression uses indexer rather than `Add()` for `KeyValuePair<K, V>` collections
+
+***Introduced in Visual Studio 2022 version 17.14***
+
+For a collection expression where the target type is a `struct` or `class` type where the *iteration type* of the target type
+is `KeyValuePair<K, V>`, and where the target type provides an indexer of the form `public V this[K key] { get; set; }`,
+the collection instance will be constructed using the indexer `set` accessor rather than applicable `Add()` methods.
+
+```csharp
+KeyValuePair<string, int> x = ...;
+IReadOnlyDictionary<string, int> y = ...;
+
+MyDictionary<string, int> z = [x, ..y]; // previously used Add(), now uses this[K].set
+
+class MyDictionary<K, V> : IEnumerable<KeyValuePair<K, V>>
+{
+    public IEnumerator<KeyValuePair<K, V>> GetEnumerator();
+    public V this[K key] { get; set; }
+    public void Add(KeyValuePair<K, V> kvp);
+}
+```
+
 ## Emitting metadata-only executables requires an entrypoint
 
 ***Introduced in Visual Studio 2022 version 17.14***
