diff --git a/src/Compilers/Core/Portable/WellKnownMember.cs b/src/Compilers/Core/Portable/WellKnownMember.cs
index c3c78f38584..dcc72a02c8e 100644
--- a/src/Compilers/Core/Portable/WellKnownMember.cs
+++ b/src/Compilers/Core/Portable/WellKnownMember.cs
@@ -718,6 +718,16 @@ internal enum WellKnownMember
         System_Text_Encoding__get_UTF8,
         System_Text_Encoding__GetString,
 
+        System_Collections_Generic_Dictionary_KV__ctor,
+        System_Collections_Generic_Dictionary_KV__ctor_IEqualityComparer_K,
+        System_Collections_Generic_Dictionary_KV__ctor_Int32,
+        System_Collections_Generic_Dictionary_KV__ctor_Int32_IEqualityComparer_K,
+        System_Collections_Generic_Dictionary_KV__set_Item,
+        System_Collections_Generic_KeyValuePair_KV__ctor,
+        System_Collections_Generic_KeyValuePair_KV__get_Key,
+        System_Collections_Generic_KeyValuePair_KV__get_Value,
+        System_Collections_ObjectModel_ReadOnlyDictionary_KV__ctor,
+
         Count,
 
         // Remember to update the AllWellKnownTypeMembers tests when making changes here
