diff --git a/src/Compilers/CSharp/Test/Emit3/Semantics/ParamsCollectionTests.cs b/src/Compilers/CSharp/Test/Emit3/Semantics/ParamsCollectionTests.cs
index 2a80d5a0d19..d047de710e3 100644
--- a/src/Compilers/CSharp/Test/Emit3/Semantics/ParamsCollectionTests.cs
+++ b/src/Compilers/CSharp/Test/Emit3/Semantics/ParamsCollectionTests.cs
@@ -943,6 +943,43 @@ internal static void Test2(params MyCollection2 a)
                 );
         }
 
+        [Fact]
+        public void CreateMethod_LocalFunction_InconsistentAccessibility()
+        {
+            string source = """
+                using System;
+                using System.Collections;
+                using System.Collections.Generic;
+                using System.Runtime.CompilerServices;
+                class Program
+                {
+                    [CollectionBuilder(typeof(MyBuilder), "Create")]
+                    private class MyCollection<T> : IEnumerable<T>
+                    {
+                        private readonly List<T> _items;
+                        public MyCollection(ReadOnlySpan<T> items) { _items = new(items.ToArray()); }
+                        public IEnumerator<T> GetEnumerator() => _items.GetEnumerator();
+                        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+                    }
+                    private class MyBuilder
+                    {
+                        public static MyCollection<T> Create<T>(ReadOnlySpan<T> items) => new(items);
+                    }
+                    static void Main()
+                    {
+                        local(1, 2, 3);
+                        void local<T>(params MyCollection<T> c) { c.Report(); }
+                    }
+                }
+                """;
+            var verifier = CompileAndVerify(
+                [source, CollectionExpressionTests.s_collectionExtensions],
+                targetFramework: TargetFramework.Net80,
+                verify: Verification.Skipped,
+                expectedOutput: ExpectedOutput("""[1, 2, 3], """));
+            verifier.VerifyDiagnostics();
+        }
+
         [Fact]
         public void CreateMethod_08_Inaccessible()
         {
@@ -2369,6 +2406,52 @@ static void Test2(params MyCollection a)
                 );
         }
 
+        [Fact]
+        public void ImplementsIEnumerableT_24_ProtectedConstructor()
+        {
+            string sourceA = """
+                using System.Collections;
+                using System.Collections.Generic;
+                internal class MyCollection<T> : IEnumerable<T>
+                {
+                    protected MyCollection() { }
+                    IEnumerator<T> IEnumerable<T>.GetEnumerator() => null;
+                    IEnumerator IEnumerable.GetEnumerator() => null;
+                    public void Add(T t) { }
+                }
+                """;
+            string sourceB = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        var f1 = (params MyCollection<string> args) => { };
+                        static void f2<T>(params MyCollection<T> args) { }
+                        f1();
+                        f2<string>();
+                    }
+                    static void F3<T>(params MyCollection<T> args) { }
+                }
+                """;
+            var comp = CreateCompilation([sourceA, sourceB]);
+            comp.VerifyEmitDiagnostics(
+                // (5,19): error CS0122: 'MyCollection<string>.MyCollection()' is inaccessible due to its protection level
+                //         var f1 = (params MyCollection<string> args) => { };
+                Diagnostic(ErrorCode.ERR_BadAccess, "params MyCollection<string> args").WithArguments("MyCollection<string>.MyCollection()").WithLocation(5, 19),
+                // (6,27): error CS0122: 'MyCollection<T>.MyCollection()' is inaccessible due to its protection level
+                //         static void f2<T>(params MyCollection<T> args) { }
+                Diagnostic(ErrorCode.ERR_BadAccess, "params MyCollection<T> args").WithArguments("MyCollection<T>.MyCollection()").WithLocation(6, 27),
+                // (7,9): error CS7036: There is no argument given that corresponds to the required parameter 'obj' of 'Action<MyCollection<string>>'
+                //         f1();
+                Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "f1").WithArguments("obj", "System.Action<MyCollection<string>>").WithLocation(7, 9),
+                // (8,9): error CS7036: There is no argument given that corresponds to the required parameter 'args' of 'f2<T>(params MyCollection<T>)'
+                //         f2<string>();
+                Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "f2<string>").WithArguments("args", "f2<T>(params MyCollection<T>)").WithLocation(8, 9),
+                // (10,23): error CS0122: 'MyCollection<T>.MyCollection()' is inaccessible due to its protection level
+                //     static void F3<T>(params MyCollection<T> args) { }
+                Diagnostic(ErrorCode.ERR_BadAccess, "params MyCollection<T> args").WithArguments("MyCollection<T>.MyCollection()").WithLocation(10, 23));
+        }
+
         [Fact]
         public void ImplementsIEnumerable_01()
         {
