diff --git a/src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs b/src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs
index 6f029786742..7d4f788ac80 100644
--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs
@@ -1581,37 +1581,41 @@ void validateParamsType(BindingDiagnosticBag diagnostics)
                                 checkIsAtLeastAsVisible(syntax, binder, constructor, diagnostics);
                             }
 
-                            if (!binder.HasCollectionExpressionApplicableAddMethod(syntax, Type, out ImmutableArray<MethodSymbol> addMethods, diagnostics))
+                            // https://github.com/dotnet/roslyn/issues/77879: Report diagnostics when GetCollectionExpressionApplicableIndexer() returns non-null?
+                            if (binder.GetCollectionExpressionApplicableIndexer(syntax, Type, elementTypeWithAnnotations.Type, BindingDiagnosticBag.Discarded) is null)
                             {
-                                return;
-                            }
-
-                            Debug.Assert(!addMethods.IsDefaultOrEmpty);
-
-                            if (addMethods[0].IsExtensionMethod || addMethods[0].GetIsNewExtensionMember()) // No need to check other methods, extensions are never mixed with instance methods
-                            {
-                                diagnostics.Add(ErrorCode.ERR_ParamsCollectionExtensionAddMethod, syntax, Type);
-                                return;
-                            }
-
-                            MethodSymbol? reportAsLessVisible = null;
-
-                            foreach (var addMethod in addMethods)
-                            {
-                                if (isAtLeastAsVisible(syntax, binder, addMethod, diagnostics))
+                                if (!binder.HasCollectionExpressionApplicableAddMethod(syntax, Type, out ImmutableArray<MethodSymbol> addMethods, diagnostics))
                                 {
-                                    reportAsLessVisible = null;
-                                    break;
+                                    return;
                                 }
-                                else
-                                {
-                                    reportAsLessVisible ??= addMethod;
-                                }
-                            }
 
-                            if (reportAsLessVisible is not null)
-                            {
-                                diagnostics.Add(ErrorCode.ERR_ParamsMemberCannotBeLessVisibleThanDeclaringMember, syntax, reportAsLessVisible, ContainingSymbol);
+                                Debug.Assert(!addMethods.IsDefaultOrEmpty);
+
+                                if (addMethods[0].IsExtensionMethod || addMethods[0].GetIsNewExtensionMember()) // No need to check other methods, extensions are never mixed with instance methods
+                                {
+                                    diagnostics.Add(ErrorCode.ERR_ParamsCollectionExtensionAddMethod, syntax, Type);
+                                    return;
+                                }
+
+                                MethodSymbol? reportAsLessVisible = null;
+
+                                foreach (var addMethod in addMethods)
+                                {
+                                    if (isAtLeastAsVisible(syntax, binder, addMethod, diagnostics))
+                                    {
+                                        reportAsLessVisible = null;
+                                        break;
+                                    }
+                                    else
+                                    {
+                                        reportAsLessVisible ??= addMethod;
+                                    }
+                                }
+
+                                if (reportAsLessVisible is not null)
+                                {
+                                    diagnostics.Add(ErrorCode.ERR_ParamsMemberCannotBeLessVisibleThanDeclaringMember, syntax, reportAsLessVisible, ContainingSymbol);
+                                }
                             }
                         }
                         break;
@@ -1629,7 +1633,12 @@ void validateParamsType(BindingDiagnosticBag diagnostics)
                                 return;
                             }
 
-                            MethodSymbol? collectionBuilderMethod = binder.GetAndValidateCollectionBuilderMethod(syntax, (NamedTypeSymbol)Type, diagnostics, elementType: out _);
+                            var targetType = (NamedTypeSymbol)Type;
+                            targetType.OriginalDefinition.HasCollectionBuilderAttribute(out TypeSymbol? builderType, out string? methodName);
+                            Debug.Assert(builderType is { });
+                            Debug.Assert(!string.IsNullOrEmpty(methodName));
+
+                            MethodSymbol? collectionBuilderMethod = binder.GetAndValidateCollectionBuilderMethod(syntax, targetType.OriginalDefinition, builderType, methodName, diagnostics);
                             if (collectionBuilderMethod is null)
                             {
                                 return;
