diff --git a/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs b/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
index f6f377cf562..9278bd1a13d 100644
--- a/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
@@ -4184,6 +4184,7 @@ internal SafeContext GetValEscape(BoundExpression expr, SafeContext localScopeDe
                 case BoundKind.DeconstructValuePlaceholder:
                 case BoundKind.InterpolatedStringArgumentPlaceholder:
                 case BoundKind.AwaitableValuePlaceholder:
+                case BoundKind.ValuePlaceholder:
                     return GetPlaceholderScope((BoundValuePlaceholderBase)expr);
 
                 case BoundKind.Local:
@@ -4634,27 +4635,13 @@ private bool HasLocalScope(BoundCollectionExpression expr)
                 case CollectionExpressionTypeKind.Span:
                     return true;
                 case CollectionExpressionTypeKind.CollectionBuilder:
+                    Debug.Assert(expr.CollectionCreation is { });
                     // For a ref struct type with a builder method, the scope of the collection
                     // expression is the scope of an invocation of the builder method with the
                     // collection expression as the span argument. That is, `R r = [x, y, z];`
                     // is equivalent to `R r = Builder.Create((ReadOnlySpan<...>)[x, y, z]);`.
-                    var constructMethod = expr.CollectionBuilderMethod;
-                    if (constructMethod is not { Parameters: [{ RefKind: RefKind.None } parameter] })
-                    {
-                        // Unexpected construct method. Restrict the collection to local scope.
-                        return true;
-                    }
-                    Debug.Assert(constructMethod.ReturnType.Equals(expr.Type, TypeCompareKind.AllIgnoreOptions));
-                    Debug.Assert(parameter.Type.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.AllIgnoreOptions));
-                    if (parameter.EffectiveScope == ScopedKind.ScopedValue)
-                    {
-                        return false;
-                    }
-                    if (LocalRewriter.ShouldUseRuntimeHelpersCreateSpan(expr, ((NamedTypeSymbol)parameter.Type).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type))
-                    {
-                        return false;
-                    }
-                    return true;
+                    var context = GetValEscape(expr.CollectionCreation, _localScopeDepth);
+                    return !context.IsReturnable;
                 case CollectionExpressionTypeKind.ImplementsIEnumerable:
                     // Error cases. Restrict the collection to local scope.
                     return true;
@@ -4876,6 +4863,7 @@ internal bool CheckValEscape(SyntaxNode node, BoundExpression expr, SafeContext
                 case BoundKind.DeconstructValuePlaceholder:
                 case BoundKind.AwaitableValuePlaceholder:
                 case BoundKind.InterpolatedStringArgumentPlaceholder:
+                case BoundKind.ValuePlaceholder:
                     if (!GetPlaceholderScope((BoundValuePlaceholderBase)expr).IsConvertibleTo(escapeTo))
                     {
                         Error(diagnostics, inUnsafeRegion ? ErrorCode.WRN_EscapeVariable : ErrorCode.ERR_EscapeVariable, node, expr.Syntax);
