diff --git a/src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs b/src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs
index 1ef95415768..7a879b815b8 100644
--- a/src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs
+++ b/src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs
@@ -202,11 +202,18 @@ CollectionExpressionSyntax CreateSingleElementCollection(CollectionMatch<TMatchN
             //
             // For that sort of case.  Single element collections should stay closely associated with the original
             // expression.
+            return CollectionExpression([CreateElement(match)]).WithTriviaFrom(expressionToReplace);
+        }
+
+        CollectionElementSyntax CreateElement(CollectionMatch<TMatchNode> match)
+        {
+            if (match.Node is ArgumentListSyntax argumentList)
+                return WithElement(argumentList.WithoutTrivia());
+
             var expression = (ExpressionSyntax)(object)match.Node;
-            return CollectionExpression([
-                match.UseSpread
-                    ? SpreadElement(expression.WithoutTrivia())
-                    : ExpressionElement(expression.WithoutTrivia())]).WithTriviaFrom(expressionToReplace);
+            return match.UseSpread
+                ? SpreadElement(expression.WithoutTrivia())
+                : ExpressionElement(expression.WithoutTrivia());
         }
 
         CollectionExpressionSyntax CreateCollectionExpressionWithExistingElements()
@@ -220,7 +227,7 @@ CollectionExpressionSyntax CreateCollectionExpressionWithExistingElements()
                 // braces (and initial elements) match whatever the initializer correct looks like.
 
                 var initialCollection = UseCollectionExpressionHelpers.ConvertInitializerToCollectionExpression(
-                    initializer, wasOnSingleLine: false);
+                    document.Text, initializer, wasOnSingleLine: false);
 
                 if (!makeMultiLineCollectionExpression &&
                     document.Text.AreOnSameLine(initializer.Expressions.First().GetFirstToken(), initializer.Expressions.Last().GetLastToken()))
@@ -259,7 +266,7 @@ CollectionExpressionSyntax CreateCollectionExpressionWithExistingElements()
                 // end.
 
                 var initialCollection = UseCollectionExpressionHelpers.ConvertInitializerToCollectionExpression(
-                    initializer, wasOnSingleLine: false);
+                    document.Text, initializer, wasOnSingleLine: false);
 
                 if (document.Text.AreOnSameLine(initializer.OpenBraceToken.GetPreviousToken(), initializer.OpenBraceToken))
                 {
@@ -308,7 +315,7 @@ CollectionExpressionSyntax CreateCollectionExpressionWithExistingElements()
                 // First, convert the existing initializer (and its expressions) into a corresponding collection
                 // expression.  This will fixup the braces properly for the collection expression.
                 var initialCollection = UseCollectionExpressionHelpers.ConvertInitializerToCollectionExpression(
-                    initializer, wasOnSingleLine: true);
+                    document.Text, initializer, wasOnSingleLine: true);
 
                 // now, add all the matches in after the existing elements.
                 var finalCollection = AddMatchesToExistingNonEmptyCollectionExpression(initialCollection, preferredIndentation: null);
@@ -461,33 +468,62 @@ IEnumerable<CollectionElementSyntax> CreateElements(
             {
                 // Create:
                 //
-                //      `x` for `collection.Add(x)`
-                //      `.. x` for `collection.AddRange(x)`
-                //      `x, y, z` for `collection.AddRange(x, y, z)`
-                var expressions = ConvertExpressions(expressionStatement.Expression, expr => IndentExpression(expressionStatement, expr, preferredIndentation));
+                //      `x: y` for `collection.Add(x, y)`               // when useKeyValue=true
+                //      `x: y` for `collection[x] = y`                  // when useKeyValue=true
+                //      `x` for `collection.Add(x)`                     // when useSpread=false
+                //      `.. x` for `collection.AddRange(x)`             // when useSpread=true
+                //      `x, y, z` for `collection.AddRange(x, y, z)`    // when useSpread=false
 
-                Contract.ThrowIfTrue(expressions.Length >= 2 && match.UseSpread);
-
-                if (match.UseSpread && expressions is [CollectionExpressionSyntax collectionExpression])
+                if (match.UseKeyValue)
                 {
-                    // If we're spreading a collection expression, just insert those inner collection expression
-                    // elements as is into the outer collection expression.
-                    foreach (var element in collectionExpression.Elements)
+                    if (expressionStatement.Expression is InvocationExpressionSyntax invocation)
                     {
-                        if (element is SpreadElementSyntax spreadElement)
-                        {
-                            yield return CreateCollectionElement(useSpread: true, spreadElement.Expression);
-                        }
-                        else if (element is ExpressionElementSyntax expressionElement)
-                        {
-                            yield return CreateCollectionElement(useSpread: false, expressionElement.Expression);
-                        }
+                        var arguments = invocation.ArgumentList.Arguments;
+                        yield return KeyValuePairElement(
+                            IndentExpression(expressionStatement, arguments[0].Expression, preferredIndentation),
+                            ColonToken.WithTriviaFrom(arguments.GetSeparator(0)),
+                            IndentExpression(expressionStatement, arguments[1].Expression, preferredIndentation));
+                    }
+                    else if (expressionStatement.Expression is AssignmentExpressionSyntax assignment)
+                    {
+                        var elementAccess = (ElementAccessExpressionSyntax)assignment.Left;
+                        yield return KeyValuePairElement(
+                            IndentExpression(expressionStatement, elementAccess.ArgumentList.Arguments[0].Expression, preferredIndentation),
+                            ColonToken.WithTrailingTrivia(assignment.OperatorToken.TrailingTrivia),
+                            IndentExpression(expressionStatement, assignment.Right, preferredIndentation));
+                    }
+                    else
+                    {
+                        throw ExceptionUtilities.Unreachable();
                     }
                 }
                 else
                 {
-                    foreach (var expression in expressions)
-                        yield return CreateCollectionElement(match.UseSpread, expression);
+                    var expressions = ConvertExpressions(expressionStatement.Expression, expr => IndentExpression(expressionStatement, expr, preferredIndentation));
+
+                    Contract.ThrowIfTrue(expressions.Length >= 2 && match.UseSpread);
+
+                    if (match.UseSpread && expressions is [CollectionExpressionSyntax collectionExpression])
+                    {
+                        // If we're spreading a collection expression, just insert those inner collection expression
+                        // elements as is into the outer collection expression.
+                        foreach (var element in collectionExpression.Elements)
+                        {
+                            if (element is SpreadElementSyntax spreadElement)
+                            {
+                                yield return CreateCollectionElement(useSpread: true, spreadElement.Expression);
+                            }
+                            else if (element is ExpressionElementSyntax expressionElement)
+                            {
+                                yield return CreateCollectionElement(useSpread: false, expressionElement.Expression);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        foreach (var expression in expressions)
+                            yield return CreateCollectionElement(match.UseSpread, expression);
+                    }
                 }
             }
             else if (node is ForEachStatementSyntax foreachStatement)
@@ -527,6 +563,10 @@ IEnumerable<CollectionElementSyntax> CreateElements(
             {
                 yield return CreateCollectionElement(match.UseSpread, IndentExpression(parentStatement: null, expression, preferredIndentation));
             }
+            else if (node is ArgumentListSyntax argumentList)
+            {
+                yield return WithElement(argumentList.WithoutTrivia());
+            }
             else
             {
                 throw ExceptionUtilities.Unreachable();
@@ -743,7 +783,7 @@ bool MakeMultiLineCollectionExpression()
 
             bool CheckForMultiLine(ImmutableArray<CollectionMatch<TMatchNode>> matches)
             {
-                foreach (var (node, _) in matches)
+                foreach (var (node, _, _) in matches)
                 {
                     // if the statement we're replacing has any comments on it, then we need to be multiline to give them an
                     // appropriate place to go.
