diff --git a/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs b/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs
index 85325ad9d36..5229d468323 100644
--- a/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs
@@ -155,6 +155,7 @@ protected override IEnumerable<TypeInferenceInfo> InferTypesWorker_DoNotCallDire
                 DoStatementSyntax doStatement => InferTypeInDoStatement(doStatement),
                 EqualsValueClauseSyntax equalsValue => InferTypeInEqualsValueClause(equalsValue),
                 ExpressionColonSyntax expressionColon => InferTypeInExpressionColon(expressionColon),
+                ExpressionElementSyntax expressionElement => InferTypeInExpressionElement(expressionElement),
                 ExpressionStatementSyntax _ => InferTypeInExpressionStatement(),
                 ForEachStatementSyntax forEachStatement => InferTypeInForEachStatement(forEachStatement, expression),
                 ForStatementSyntax forStatement => InferTypeInForStatement(forStatement, expression),
@@ -162,6 +163,7 @@ protected override IEnumerable<TypeInferenceInfo> InferTypesWorker_DoNotCallDire
                 InitializerExpressionSyntax initializerExpression => InferTypeInInitializerExpression(initializerExpression, expression),
                 IsPatternExpressionSyntax isPatternExpression => InferTypeInIsPatternExpression(isPatternExpression, node),
                 LockStatementSyntax lockStatement => InferTypeInLockStatement(lockStatement),
+                KeyValuePairElementSyntax keyValuePairElement => InferTypeInKeyValuePairElement(keyValuePairElement, expression),
                 MemberAccessExpressionSyntax memberAccessExpression => InferTypeInMemberAccessExpression(memberAccessExpression, expression),
                 NameColonSyntax nameColon => InferTypeInNameColon(nameColon),
                 NameEqualsSyntax nameEquals => InferTypeInNameEquals(nameEquals),
@@ -1233,6 +1235,74 @@ private IEnumerable<TypeInferenceInfo> InferTypeInPropertyDeclaration(PropertyDe
             return CreateResult(typeInfo.Type);
         }
 
+        private IEnumerable<TypeInferenceInfo> InferTypeInExpressionElement(ExpressionElementSyntax expressionElement)
+        {
+            if (expressionElement.Parent is CollectionExpressionSyntax collectionExpression)
+            {
+                var collectionType = SemanticModel.GetTypeInfo(collectionExpression, CancellationToken).ConvertedType;
+
+                // Try to figure out the type based on the type of the collection itself.
+                if (CollectionExpressionUtilities.IsConstructibleCollectionType(
+                        SemanticModel.Compilation, collectionType, out var elementType))
+                {
+                    return [new(elementType)];
+                }
+
+                // If that fails, see if we can figure out from one of our siblings.
+                foreach (var element in collectionExpression.Elements)
+                {
+                    if (element != expressionElement && element is ExpressionElementSyntax siblingElement)
+                    {
+                        var types = GetTypes(siblingElement.Expression, objectAsDefault: false);
+                        if (types.Any())
+                            return types;
+                    }
+                }
+            }
+
+            return [];
+        }
+
+        private IEnumerable<TypeInferenceInfo> InferTypeInKeyValuePairElement(
+            KeyValuePairElementSyntax keyValuePairElement, ExpressionSyntax expression)
+        {
+            var isKey = expression == keyValuePairElement.KeyExpression;
+            var types = InferTypes();
+            return types.Select(t => new TypeInferenceInfo(t));
+
+            IEnumerable<ITypeSymbol> InferTypes()
+            {
+                if (keyValuePairElement.Parent is CollectionExpressionSyntax collectionExpression)
+                {
+                    var collectionType = SemanticModel.GetTypeInfo(collectionExpression, CancellationToken).ConvertedType;
+
+                    // Try to figure out the type based on the type of hte collection itself.
+                    if (CollectionExpressionUtilities.IsConstructibleCollectionType(SemanticModel.Compilation, collectionType, out var elementType) &&
+                        elementType is INamedTypeSymbol
+                        {
+                            Name: nameof(KeyValuePair<int, int>),
+                            TypeArguments: [var keyType, var valueType]
+                        })
+                    {
+                        return [isKey ? keyType : valueType];
+                    }
+
+                    // If that fails, see if we can figure out from one of our siblings.
+                    foreach (var element in collectionExpression.Elements)
+                    {
+                        if (element != keyValuePairElement && element is KeyValuePairElementSyntax siblingElement)
+                        {
+                            var types = GetTypes(isKey ? siblingElement.KeyExpression : siblingElement.ValueExpression, objectAsDefault: false);
+                            if (types.Any())
+                                return types.Select(t => t.InferredType);
+                        }
+                    }
+                }
+
+                return [];
+            }
+        }
+
         private IEnumerable<TypeInferenceInfo> InferTypeInExpressionStatement(SyntaxToken? previousToken = null)
         {
             // If we're position based, then that means we're after the semicolon.  In this case
