diff --git a/src/Compilers/Core/Portable/MemberDescriptor.cs b/src/Compilers/Core/Portable/MemberDescriptor.cs
index b9d932f0728..3bcc3226bba 100644
--- a/src/Compilers/Core/Portable/MemberDescriptor.cs
+++ b/src/Compilers/Core/Portable/MemberDescriptor.cs
@@ -19,13 +19,14 @@ internal enum MemberFlags : byte
         Field = 0x02,
         Constructor = 0x04,
         PropertyGet = 0x08,
-        Property = 0x10,
+        PropertySet = 0x10,
+        Property = 0x20,
         // END Mutually exclusive Member kinds
 
-        KindMask = 0x1F,
+        KindMask = 0x3F,
 
-        Static = 0x20,
-        Virtual = 0x40, // Virtual in CLR terms, i.e. sealed should be accepted.
+        Static = 0x40,
+        Virtual = 0x80, // Virtual in CLR terms, i.e. sealed should be accepted.
     }
 
     /// <summary>
