diff --git a/src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests.cs b/src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests.cs
index 45dd91e11bc..0be91c1fabf 100644
--- a/src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests.cs
+++ b/src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests.cs
@@ -4531,15 +4531,15 @@ static void Main()
                 """;
             var comp = CreateCompilation(source);
             comp.VerifyEmitDiagnostics(
-                // (9,31): warning CS8601: Possible null reference assignment.
+                // (9,29): warning CS8601: Possible null reference assignment.
                 //         object[] ab = [..a, ..b]; // 1
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "b").WithLocation(9, 31),
-                // (10,26): warning CS8601: Possible null reference assignment.
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "..b").WithLocation(9, 29),
+                // (10,24): warning CS8601: Possible null reference assignment.
                 //         object[] bb = [..b, ..b]; // 2
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "b").WithLocation(10, 26),
-                // (10,31): warning CS8601: Possible null reference assignment.
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "..b").WithLocation(10, 24),
+                // (10,29): warning CS8601: Possible null reference assignment.
                 //         object[] bb = [..b, ..b]; // 2
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "b").WithLocation(10, 31));
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "..b").WithLocation(10, 29));
         }
 
         [Fact]
@@ -4562,15 +4562,15 @@ static void Main()
                 """;
             var comp = CreateCompilation(source);
             comp.VerifyEmitDiagnostics(
-                // (10,44): warning CS8619: Nullability of reference types in value of type 'IEnumerable<string?>' doesn't match target type 'IEnumerable<object>'.
+                // (10,42): warning CS8619: Nullability of reference types in value of type 'IEnumerable<string?>' doesn't match target type 'IEnumerable<object>'.
                 //         IEnumerable<object>[] ab = [..a, ..b]; // 1
-                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "b").WithArguments("System.Collections.Generic.IEnumerable<string?>", "System.Collections.Generic.IEnumerable<object>").WithLocation(10, 44),
-                // (11,39): warning CS8619: Nullability of reference types in value of type 'IEnumerable<string?>' doesn't match target type 'IEnumerable<object>'.
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "..b").WithArguments("System.Collections.Generic.IEnumerable<string?>", "System.Collections.Generic.IEnumerable<object>").WithLocation(10, 42),
+                // (11,37): warning CS8619: Nullability of reference types in value of type 'IEnumerable<string?>' doesn't match target type 'IEnumerable<object>'.
                 //         IEnumerable<object>[] bb = [..b, ..b]; // 2
-                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "b").WithArguments("System.Collections.Generic.IEnumerable<string?>", "System.Collections.Generic.IEnumerable<object>").WithLocation(11, 39),
-                // (11,44): warning CS8619: Nullability of reference types in value of type 'IEnumerable<string?>' doesn't match target type 'IEnumerable<object>'.
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "..b").WithArguments("System.Collections.Generic.IEnumerable<string?>", "System.Collections.Generic.IEnumerable<object>").WithLocation(11, 37),
+                // (11,42): warning CS8619: Nullability of reference types in value of type 'IEnumerable<string?>' doesn't match target type 'IEnumerable<object>'.
                 //         IEnumerable<object>[] bb = [..b, ..b]; // 2
-                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "b").WithArguments("System.Collections.Generic.IEnumerable<string?>", "System.Collections.Generic.IEnumerable<object>").WithLocation(11, 44));
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "..b").WithArguments("System.Collections.Generic.IEnumerable<string?>", "System.Collections.Generic.IEnumerable<object>").WithLocation(11, 42));
         }
 
         [Fact]
@@ -5123,31 +5123,56 @@ static void F(IEnumerable<int> e)
                 """;
             var comp = CreateCompilation(source);
             comp.MakeTypeMissing(WellKnownType.System_Collections_Generic_List_T);
-            comp.VerifyEmitDiagnostics(
-                // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
-                //         c = [];
-                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(7, 13),
-                // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
-                //         c = [];
-                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(7, 13),
-                // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.Add'
-                //         c = [];
-                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", "Add").WithLocation(7, 13),
-                // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.ToArray'
-                //         c = [];
-                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", "ToArray").WithLocation(7, 13),
-                // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
-                //         c = [..e];
-                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(8, 13),
-                // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
-                //         c = [..e];
-                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(8, 13),
-                // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.Add'
-                //         c = [..e];
-                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", "Add").WithLocation(8, 13),
-                // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.ToArray'
-                //         c = [..e];
-                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", "ToArray").WithLocation(8, 13));
+            if (collectionType is "ICollection<object>" or "IList<int>")
+            {
+                comp.VerifyEmitDiagnostics(
+                    // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
+                    //         c = [];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(7, 13),
+                    // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
+                    //         c = [];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(7, 13),
+                    // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.Add'
+                    //         c = [];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", "Add").WithLocation(7, 13),
+                    // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.ToArray'
+                    //         c = [];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", "ToArray").WithLocation(7, 13),
+                    // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
+                    //         c = [..e];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(8, 13),
+                    // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
+                    //         c = [..e];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(8, 13),
+                    // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.Add'
+                    //         c = [..e];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", "Add").WithLocation(8, 13),
+                    // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.ToArray'
+                    //         c = [..e];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", "ToArray").WithLocation(8, 13));
+            }
+            else
+            {
+                comp.VerifyEmitDiagnostics(
+                    // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.Add'
+                    //         c = [];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", "Add").WithLocation(7, 13),
+                    // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.ToArray'
+                    //         c = [];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", "ToArray").WithLocation(7, 13),
+                    // (7,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
+                    //         c = [];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(7, 13),
+                    // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.Add'
+                    //         c = [..e];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", "Add").WithLocation(8, 13),
+                    // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.ToArray'
+                    //         c = [..e];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", "ToArray").WithLocation(8, 13),
+                    // (8,13): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
+                    //         c = [..e];
+                    Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..e]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(8, 13));
+            }
         }
 
         [Fact]
@@ -5695,7 +5720,7 @@ static void Main()
         }
 
         [Fact]
-        public void InterfaceType()
+        public void InterfaceType_ComImport()
         {
             string source = """
                 using System;
@@ -5738,6 +5763,94 @@ static void Main()
                 Diagnostic(ErrorCode.ERR_CollectionExpressionTargetTypeNotConstructible, "[3, 4]").WithArguments("I").WithLocation(27, 13));
         }
 
+        [Fact]
+        public void InterfaceType_ImplementsIEnumerable_01()
+        {
+            string source = $$"""
+                using System.Collections;
+                using System.Collections.Generic;
+                interface IMyCollection1<T> : IEnumerable
+                {
+                    void Add(T t);
+                }
+                interface IMyCollection2<T> : IEnumerable<T>
+                {
+                    void Add(T t);
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        IMyCollection1<int> x;
+                        x = [];
+                        x = [1];
+                        IMyCollection2<int> y;
+                        y = [];
+                        y = [2];
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (16,13): error CS9174: Cannot initialize type 'IMyCollection1<int>' with a collection expression because the type is not constructible.
+                //         x = [];
+                Diagnostic(ErrorCode.ERR_CollectionExpressionTargetTypeNotConstructible, "[]").WithArguments("IMyCollection1<int>").WithLocation(16, 13),
+                // (17,13): error CS9174: Cannot initialize type 'IMyCollection1<int>' with a collection expression because the type is not constructible.
+                //         x = [1];
+                Diagnostic(ErrorCode.ERR_CollectionExpressionTargetTypeNotConstructible, "[1]").WithArguments("IMyCollection1<int>").WithLocation(17, 13),
+                // (19,13): error CS9174: Cannot initialize type 'IMyCollection2<int>' with a collection expression because the type is not constructible.
+                //         y = [];
+                Diagnostic(ErrorCode.ERR_CollectionExpressionTargetTypeNotConstructible, "[]").WithArguments("IMyCollection2<int>").WithLocation(19, 13),
+                // (20,13): error CS9174: Cannot initialize type 'IMyCollection2<int>' with a collection expression because the type is not constructible.
+                //         y = [2];
+                Diagnostic(ErrorCode.ERR_CollectionExpressionTargetTypeNotConstructible, "[2]").WithArguments("IMyCollection2<int>").WithLocation(20, 13));
+        }
+
+        [Fact]
+        public void InterfaceType_ImplementsIEnumerable_02()
+        {
+            string source = $$"""
+                using System.Collections;
+                using System.Collections.Generic;
+                interface IMyCollection1<T> : IEnumerable
+                {
+                    void Add(T t) { }
+                }
+                interface IMyCollection2<T> : IEnumerable<T>
+                {
+                    void Add(T t) { }
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        IMyCollection1<int> x;
+                        x = [];
+                        x = [1];
+                        IMyCollection2<int> y;
+                        y = [];
+                        y = [2];
+                    }
+                }
+                """;
+            var comp = CreateCompilation(
+                source,
+                targetFramework: TargetFramework.Net80);
+            comp.VerifyEmitDiagnostics(
+                // (16,13): error CS9174: Cannot initialize type 'IMyCollection1<int>' with a collection expression because the type is not constructible.
+                //         x = [];
+                Diagnostic(ErrorCode.ERR_CollectionExpressionTargetTypeNotConstructible, "[]").WithArguments("IMyCollection1<int>").WithLocation(16, 13),
+                // (17,13): error CS9174: Cannot initialize type 'IMyCollection1<int>' with a collection expression because the type is not constructible.
+                //         x = [1];
+                Diagnostic(ErrorCode.ERR_CollectionExpressionTargetTypeNotConstructible, "[1]").WithArguments("IMyCollection1<int>").WithLocation(17, 13),
+                // (19,13): error CS9174: Cannot initialize type 'IMyCollection2<int>' with a collection expression because the type is not constructible.
+                //         y = [];
+                Diagnostic(ErrorCode.ERR_CollectionExpressionTargetTypeNotConstructible, "[]").WithArguments("IMyCollection2<int>").WithLocation(19, 13),
+                // (20,13): error CS9174: Cannot initialize type 'IMyCollection2<int>' with a collection expression because the type is not constructible.
+                //         y = [2];
+                Diagnostic(ErrorCode.ERR_CollectionExpressionTargetTypeNotConstructible, "[2]").WithArguments("IMyCollection2<int>").WithLocation(20, 13));
+        }
+
         [Fact]
         public void StringType_01()
         {
@@ -9342,25 +9455,42 @@ static void Main()
                     {
                         Dictionary<int, int> d;
                         d = [default];
+                        d.Report();
                         d = [new KeyValuePair<int, int>(1, 2)];
+                        d.Report();
+                        d = [3:4];
+                        d.Report();
+                    }
+                }
+                """;
+            CompileAndVerify(new[] { source, s_collectionExtensions }, expectedOutput: "[[0, 0]], [[1, 2]], [[3, 4]], ");
+        }
+
+        [Fact]
+        public void KeyValuePairElement_CSharp13()
+        {
+            string source = """
+                using System.Collections.Generic;
+                class Program
+                {
+                    static void Main()
+                    {
+                        Dictionary<int, int> d;
                         d = [3:4];
                     }
                 }
                 """;
-            var comp = CreateCompilation(source);
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular13);
             comp.VerifyEmitDiagnostics(
                 // (7,13): error CS9215: Collection expression type 'Dictionary<int, int>' must have an instance or extension method 'Add' that can be called with a single argument.
-                //         d = [default];
-                Diagnostic(ErrorCode.ERR_CollectionExpressionMissingAdd, "[default]").WithArguments("System.Collections.Generic.Dictionary<int, int>").WithLocation(7, 13),
-                // (8,13): error CS9215: Collection expression type 'Dictionary<int, int>' must have an instance or extension method 'Add' that can be called with a single argument.
-                //         d = [new KeyValuePair<int, int>(1, 2)];
-                Diagnostic(ErrorCode.ERR_CollectionExpressionMissingAdd, "[new KeyValuePair<int, int>(1, 2)]").WithArguments("System.Collections.Generic.Dictionary<int, int>").WithLocation(8, 13),
-                // (9,15): error CS1003: Syntax error, ',' expected
                 //         d = [3:4];
-                Diagnostic(ErrorCode.ERR_SyntaxError, ":").WithArguments(",").WithLocation(9, 15),
-                // (9,16): error CS1003: Syntax error, ',' expected
+                Diagnostic(ErrorCode.ERR_CollectionExpressionMissingAdd, "[3:4]").WithArguments("System.Collections.Generic.Dictionary<int, int>").WithLocation(7, 13),
+                // (7,14): error CS9500: Collection expression type 'Dictionary<int, int>' does not support key-value pair elements.
                 //         d = [3:4];
-                Diagnostic(ErrorCode.ERR_SyntaxError, "4").WithArguments(",").WithLocation(9, 16));
+                Diagnostic(ErrorCode.ERR_CollectionExpressionKeyValuePairNotSupported, "3:4").WithArguments("System.Collections.Generic.Dictionary<int, int>").WithLocation(7, 14),
+                // (7,15): error CS8652: The feature 'dictionary expressions' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         d = [3:4];
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, ":").WithArguments("dictionary expressions").WithLocation(7, 15));
         }
 
         [Theory]
@@ -13814,9 +13944,6 @@ static void Main()
             var comp = CreateCompilation(source);
             comp.MakeTypeMissing(WellKnownType.System_Collections_Generic_List_T);
             comp.VerifyEmitDiagnostics(
-                // (6,30): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
-                //         IEnumerable<int> x = [0];
-                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[0]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(6, 30),
                 // (6,30): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
                 //         IEnumerable<int> x = [0];
                 Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[0]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(6, 30),
@@ -13829,9 +13956,6 @@ static void Main()
                 // (7,30): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
                 //         IEnumerable<int> y = [..x];
                 Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..x]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(7, 30),
-                // (7,30): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1..ctor'
-                //         IEnumerable<int> y = [..x];
-                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..x]").WithArguments("System.Collections.Generic.List`1", ".ctor").WithLocation(7, 30),
                 // (7,30): error CS0656: Missing compiler required member 'System.Collections.Generic.List`1.Add'
                 //         IEnumerable<int> y = [..x];
                 Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "[..x]").WithArguments("System.Collections.Generic.List`1", "Add").WithLocation(7, 30),
@@ -14128,9 +14252,9 @@ public void Nullable_Spread_01()
                 }
                 """;
             CreateCompilation(source).VerifyDiagnostics(
-                // (5,21): warning CS8601: Possible null reference assignment.
+                // (5,19): warning CS8601: Possible null reference assignment.
                 //     object[] b = [..a];
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "a").WithLocation(5, 21));
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "..a").WithLocation(5, 19));
         }
 
         [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75560")]
@@ -14173,9 +14297,28 @@ public void Nullable_Spread_03()
                 bool m() => throw null!;
                 """;
             CreateCompilation(source).VerifyDiagnostics(
-                // (6,22): warning CS8601: Possible null reference assignment.
+                // (6,19): warning CS8601: Possible null reference assignment.
                 //     object[] b = [..(m() ? a1 : a2)];
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "m() ? a1 : a2").WithLocation(6, 22));
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "..(m() ? a1 : a2)").WithLocation(6, 19));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75560")]
+        public void Nullable_Spread_04()
+        {
+            var source = """
+                #nullable enable
+                using System.Collections.Generic;
+                {
+                    IEnumerable<object?> a = [null];
+                    List<object> b = [..a];
+                }
+                {
+                    IEnumerable<object?> a = [null];
+                    List<object> b = [..a!];
+                }
+                """;
+            // https://github.com/dotnet/roslyn/issues/68786: Should report warning for [..a].
+            CreateCompilation(source).VerifyDiagnostics();
         }
 
         [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/69447")]
@@ -17893,11 +18036,8 @@ public void CollectionBuilder_UnexpectedSignature_01(
                 "public static MyCollection<int> Create(ReadOnlySpan<int> items) => default;", // constructed parameter and return types
                 "public static MyCollection<T> Create<T>(ReadOnlySpan<int> items) => default;", // constructed parameter type
                 "public static MyCollection<int> Create<T>(ReadOnlySpan<T> items) => default;", // constructed return type
-                "public static MyCollection<T> Create<T>(ReadOnlySpan<T> items, int index = 0) => default;", // optional parameter
                 "public static MyCollection<T> Create<T>() => default;", // no parameters
                 "public static void Create<T>(ReadOnlySpan<T> items) { }", // no return type
-                "public static MyCollection<T> Create<T>(ReadOnlySpan<T> items, int index = 0) => default;", // optional parameter
-                "public static MyCollection<T> Create<T>(ReadOnlySpan<T> items, params object[] args) => default;", // params
                 "public static MyCollection<T> Create<T, U>(ReadOnlySpan<T> items) => default;", // extra type parameter
                 "public static MyCollection<T> Create<T>(Span<T> items) => default;", // Span<T>
                 "public static MyCollection<T> Create<T>(T[] items) => default;", // T[]
@@ -17948,6 +18088,112 @@ static void Main()
                 Diagnostic(ErrorCode.ERR_CollectionBuilderAttributeMethodNotFound, "[1, 2, 3]").WithArguments("Create", "T", "MyCollection<T>").WithLocation(7, 31));
         }
 
+        [CombinatorialData]
+        [Theory]
+        public void CollectionBuilder_UnexpectedSignature_02(
+            [CombinatorialValues(
+                "public static MyCollection<T> Create<T>(ReadOnlySpan<T> items, int index = 0) => new(items);", // optional parameter
+                "public static MyCollection<T> Create<T>(ReadOnlySpan<T> items, params object[] args) => new(items);")] // params
+            string methodDeclaration,
+            bool useCompilationReference)
+        {
+            string sourceA = $$"""
+                using System;
+                using System.Collections;
+                using System.Collections.Generic;
+                using System.Runtime.CompilerServices;
+                [CollectionBuilder(typeof(MyCollectionBuilder), "Create")]
+                public struct MyCollection<T> : IEnumerable<T>
+                {
+                    private readonly List<T> _list;
+                    internal MyCollection(ReadOnlySpan<T> items) { _list = new(items.ToArray()); }
+                    IEnumerator<T> IEnumerable<T>.GetEnumerator() => _list.GetEnumerator();
+                    IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
+                }
+                public class MyCollectionBuilder
+                {
+                    {{methodDeclaration}}
+                }
+                """;
+            var comp = CreateCompilation(sourceA, targetFramework: TargetFramework.Net80);
+            var refA = AsReference(comp, useCompilationReference);
+
+            string sourceB = """
+                #pragma warning disable 219
+                class Program
+                {
+                    static void Main()
+                    {
+                        MyCollection<string> x = [];
+                        x.Report();
+                        MyCollection<int> y = [1, 2, 3];
+                        y.Report();
+                        MyCollection<object> z = new();
+                    }
+                }
+                """;
+            comp = CreateCompilation([sourceB, s_collectionExtensions], references: new[] { refA }, targetFramework: TargetFramework.Net80, options: TestOptions.ReleaseExe);
+            comp.VerifyEmitDiagnostics(
+                // (6,34): error CS9187: Could not find an accessible 'Create' method with the expected signature: a static method with a single parameter of type 'ReadOnlySpan<T>' and return type 'MyCollection<T>'.
+                //         MyCollection<string> x = [];
+                Diagnostic(ErrorCode.ERR_CollectionBuilderAttributeMethodNotFound, "[]").WithArguments("Create", "T", "MyCollection<T>").WithLocation(6, 34),
+                // (8,31): error CS9187: Could not find an accessible 'Create' method with the expected signature: a static method with a single parameter of type 'ReadOnlySpan<T>' and return type 'MyCollection<T>'.
+                //         MyCollection<int> y = [1, 2, 3];
+                Diagnostic(ErrorCode.ERR_CollectionBuilderAttributeMethodNotFound, "[1, 2, 3]").WithArguments("Create", "T", "MyCollection<T>").WithLocation(8, 31));
+        }
+
+        [CombinatorialData]
+        [Theory]
+        public void CollectionBuilder_UnexpectedSignature_03(
+            bool useCompilationReference)
+        {
+            string sourceA = $$"""
+                using System;
+                using System.Collections;
+                using System.Collections.Generic;
+                using System.Runtime.CompilerServices;
+                [CollectionBuilder(typeof(MyCollectionBuilder), "Create")]
+                public struct MyCollection : IEnumerable<int>
+                {
+                    private readonly List<int> _list;
+                    internal MyCollection(ReadOnlySpan<int> items) { _list = new(items.ToArray()); }
+                    IEnumerator<int> IEnumerable<int>.GetEnumerator() => _list.GetEnumerator();
+                    IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
+                }
+                public class MyCollectionBuilder
+                {
+                    private static MyCollection _c;
+                    public static ref MyCollection Create(ReadOnlySpan<int> items)
+                    {
+                        _c = new(items);
+                        return ref _c;
+                    }
+                }
+                """;
+            var comp = CreateCompilation(sourceA, targetFramework: TargetFramework.Net80);
+            var refA = AsReference(comp, useCompilationReference);
+
+            string sourceB = """
+                #pragma warning disable 219
+                class Program
+                {
+                    static void Main()
+                    {
+                        MyCollection x = [];
+                        x.Report();
+                        MyCollection y = [1, 2, 3];
+                        y.Report();
+                    }
+                }
+                """;
+            comp = CreateCompilation([sourceB, s_collectionExtensions], references: new[] { refA }, targetFramework: TargetFramework.Net80, options: TestOptions.ReleaseExe);
+            comp.VerifyEmitDiagnostics();
+            CompileAndVerify(
+                comp,
+                verify: Verification.Skipped,
+                expectedOutput: IncludeExpectedOutput("[], [1, 2, 3], "));
+        }
+
         [CombinatorialData]
         [Theory]
         public void CollectionBuilder_UnexpectedSignature_MoreTypeParameters(bool useCompilationReference)
@@ -18860,7 +19106,9 @@ static void Main()
                         MyCollection<string> x = [];
                         MyCollection<int> y = [1, 2, 3];
                         MyCollection<object> z = new();
+                        y = Params(4, 5);
                     }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
                 }
                 """;
             comp = CreateCompilation(sourceB, references: new[] { refA }, targetFramework: TargetFramework.Net80);
@@ -18870,7 +19118,13 @@ static void Main()
                 Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "[]").WithArguments("MyCollectionBuilder").WithLocation(6, 34),
                 // (7,31): warning CS0612: 'MyCollectionBuilder' is obsolete
                 //         MyCollection<int> y = [1, 2, 3];
-                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "[1, 2, 3]").WithArguments("MyCollectionBuilder").WithLocation(7, 31));
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "[1, 2, 3]").WithArguments("MyCollectionBuilder").WithLocation(7, 31),
+                // (9,13): warning CS0612: 'MyCollectionBuilder' is obsolete
+                //         y = Params(4, 5);
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "Params(4, 5)").WithArguments("MyCollectionBuilder").WithLocation(9, 13),
+                // (11,38): warning CS0612: 'MyCollectionBuilder' is obsolete
+                //     static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "params MyCollection<T> c").WithArguments("MyCollectionBuilder").WithLocation(11, 38));
         }
 
         [Fact]
@@ -18902,7 +19156,9 @@ static void Main()
                         MyCollection<string> x = [];
                         MyCollection<int> y = [1, 2, 3];
                         MyCollection<object> z = new();
+                        y = Params(4, 5);
                     }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
                 }
                 """;
             var comp = CreateCompilation(new[] { sourceA, sourceB }, targetFramework: TargetFramework.Net80);
@@ -18915,7 +19171,13 @@ static void Main()
                 Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "[]").WithArguments("MyCollectionBuilder", "message 2").WithLocation(6, 34),
                 // 1.cs(7,31): error CS0619: 'MyCollectionBuilder' is obsolete: 'message 2'
                 //         MyCollection<int> y = [1, 2, 3];
-                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "[1, 2, 3]").WithArguments("MyCollectionBuilder", "message 2").WithLocation(7, 31));
+                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "[1, 2, 3]").WithArguments("MyCollectionBuilder", "message 2").WithLocation(7, 31),
+                // 1.cs(9,13): error CS0619: 'MyCollectionBuilder' is obsolete: 'message 2'
+                //         y = Params(4, 5);
+                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "Params(4, 5)").WithArguments("MyCollectionBuilder", "message 2").WithLocation(9, 13),
+                // 1.cs(11,38): error CS0619: 'MyCollectionBuilder' is obsolete: 'message 2'
+                //     static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
+                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "params MyCollection<T> c").WithArguments("MyCollectionBuilder", "message 2").WithLocation(11, 38));
         }
 
         [CombinatorialData]
@@ -18951,7 +19213,9 @@ static void Main()
                         MyCollection<string> x = [];
                         MyCollection<int> y = [1, 2, 3];
                         MyCollection<object> z = new();
+                        y = Params(4, 5);
                     }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
                 }
                 """;
             comp = CreateCompilation(sourceB, references: new[] { refA }, targetFramework: TargetFramework.Net80);
@@ -18961,7 +19225,10 @@ static void Main()
                 Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "[]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)").WithLocation(6, 34),
                 // (7,31): warning CS0612: 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)' is obsolete
                 //         MyCollection<int> y = [1, 2, 3];
-                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "[1, 2, 3]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)").WithLocation(7, 31));
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "[1, 2, 3]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)").WithLocation(7, 31),
+                // (9,13): warning CS0612: 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)' is obsolete
+                //         y = Params(4, 5);
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "Params(4, 5)").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)").WithLocation(9, 13));
         }
 
         [CombinatorialData]
@@ -18997,7 +19264,9 @@ static void Main()
                         MyCollection<string> x = [];
                         MyCollection<int> y = [1, 2, 3];
                         MyCollection<object> z = new();
+                        y = Params(4, 5);
                     }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
                 }
                 """;
             comp = CreateCompilation(sourceB, references: new[] { refA }, targetFramework: TargetFramework.Net80);
@@ -19007,7 +19276,10 @@ static void Main()
                 Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "[]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "message 4").WithLocation(6, 34),
                 // (7,31): error CS0619: 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)' is obsolete: 'message 4'
                 //         MyCollection<int> y = [1, 2, 3];
-                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "[1, 2, 3]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "message 4").WithLocation(7, 31));
+                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "[1, 2, 3]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "message 4").WithLocation(7, 31),
+                // (9,13): error CS0619: 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)' is obsolete: 'message 4'
+                //         y = Params(4, 5);
+                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "Params(4, 5)").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "message 4").WithLocation(9, 13));
         }
 
         [Fact]
@@ -19040,7 +19312,9 @@ static void Main()
                         MyCollection<string> x = [];
                         MyCollection<int> y = [1, 2, 3];
                         MyCollection<object> z = new();
+                        y = Params(4, 5);
                     }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
                 }
                 """;
             var comp = CreateCompilation(new[] { sourceA, sourceB }, targetFramework: TargetFramework.Net80);
@@ -19051,6 +19325,9 @@ static void Main()
                 // 1.cs(7,31): error CS8901: 'MyCollectionBuilder.Create<int>(ReadOnlySpan<int>)' is attributed with 'UnmanagedCallersOnly' and cannot be called directly. Obtain a function pointer to this method.
                 //         MyCollection<int> y = [1, 2, 3];
                 Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "[1, 2, 3]").WithArguments("MyCollectionBuilder.Create<int>(System.ReadOnlySpan<int>)").WithLocation(7, 31),
+                // 1.cs(9,13): error CS8901: 'MyCollectionBuilder.Create<int>(ReadOnlySpan<int>)' is attributed with 'UnmanagedCallersOnly' and cannot be called directly. Obtain a function pointer to this method.
+                //         y = Params(4, 5);
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "Params(4, 5)").WithArguments("MyCollectionBuilder.Create<int>(System.ReadOnlySpan<int>)").WithLocation(9, 13),
                 // 0.cs(14,6): error CS8895: Methods attributed with 'UnmanagedCallersOnly' cannot have generic type parameters and cannot be declared in a generic type.
                 //     [UnmanagedCallersOnly]
                 Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodOrTypeCannotBeGeneric, "UnmanagedCallersOnly").WithLocation(14, 6),
@@ -19097,12 +19374,13 @@ static void Main()
                     }
                 }
                 """;
-            CompileAndVerify(
+            var verifier = CompileAndVerify(
                 new[] { sourceB1, s_collectionExtensions },
                 references: new[] { refA },
                 targetFramework: TargetFramework.Net80,
                 verify: Verification.FailsPEVerify,
                 expectedOutput: IncludeExpectedOutput("[], [1, 2, 3], "));
+            verifier.VerifyDiagnostics();
 
             string sourceB2 = """
                 #pragma warning disable 219
@@ -19120,6 +19398,46 @@ static void Main()
                 // (6,32): error CS0453: The type 'int?' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)'
                 //         MyCollection<int?> x = [4, null];
                 Diagnostic(ErrorCode.ERR_ValConstraintNotSatisfied, "[4, null]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "T", "int?").WithLocation(6, 32));
+
+            string sourceB3 = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        Params(4, 5).Report();
+                    }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) where T : struct => c;
+                }
+                """;
+            verifier = CompileAndVerify(
+                [sourceB3, s_collectionExtensions],
+                references: new[] { refA },
+                targetFramework: TargetFramework.Net80,
+                verify: Verification.Skipped,
+                expectedOutput: IncludeExpectedOutput("[4, 5], "));
+            verifier.VerifyDiagnostics();
+
+            string sourceB4 = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        Params<object>(4, 5);
+                    }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
+                }
+                """;
+            comp = CreateCompilation(sourceB4, references: new[] { refA }, targetFramework: TargetFramework.Net80);
+            comp.VerifyEmitDiagnostics(
+                // (5,9): error CS0453: The type 'object' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)'
+                //         Params<object>(4, 5);
+                Diagnostic(ErrorCode.ERR_ValConstraintNotSatisfied, "Params<object>(4, 5)").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "T", "object").WithLocation(5, 9),
+                // (7,28): error CS0310: 'T' must be a non-abstract type with a public parameterless constructor in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
+                //     static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
+                Diagnostic(ErrorCode.ERR_NewConstraintNotSatisfied, "Params").WithArguments("MyCollection<T>", "T", "T").WithLocation(7, 28),
+                // (7,61): error CS0310: 'T' must be a non-abstract type with a public parameterless constructor in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
+                //     static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
+                Diagnostic(ErrorCode.ERR_NewConstraintNotSatisfied, "c").WithArguments("MyCollection<T>", "T", "T").WithLocation(7, 61));
         }
 
         [CombinatorialData]
@@ -19160,12 +19478,13 @@ static void Main()
                     }
                 }
                 """;
-            CompileAndVerify(
+            var verifier = CompileAndVerify(
                 new[] { sourceB1, s_collectionExtensions },
                 references: new[] { refA },
                 targetFramework: TargetFramework.Net80,
                 verify: Verification.FailsPEVerify,
                 expectedOutput: IncludeExpectedOutput("[], [1, 2, 3], "));
+            verifier.VerifyDiagnostics();
 
             string sourceB2 = """
                 #pragma warning disable 219
@@ -19183,6 +19502,40 @@ static void Main()
                 // (6,32): error CS0453: The type 'int?' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)'
                 //         MyCollection<int?> x = [4, null];
                 Diagnostic(ErrorCode.ERR_ValConstraintNotSatisfied, "[4, null]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "T", "int?").WithLocation(6, 32));
+
+            string sourceB3 = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        Params(4, 5).Report();
+                    }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) where T : struct => c;
+                }
+                """;
+            verifier = CompileAndVerify(
+                [sourceB3, s_collectionExtensions],
+                references: new[] { refA },
+                targetFramework: TargetFramework.Net80,
+                verify: Verification.Skipped,
+                expectedOutput: IncludeExpectedOutput("[4, 5], "));
+            verifier.VerifyDiagnostics();
+
+            string sourceB4 = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        Params<object>(4, 5);
+                    }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
+                }
+                """;
+            comp = CreateCompilation(sourceB4, references: new[] { refA }, targetFramework: TargetFramework.Net80);
+            comp.VerifyEmitDiagnostics(
+                // (5,9): error CS0453: The type 'object' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)'
+                //         Params<object>(4, 5);
+                Diagnostic(ErrorCode.ERR_ValConstraintNotSatisfied, "Params<object>(4, 5)").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "T", "object").WithLocation(5, 9));
         }
 
         [Fact]
@@ -19207,7 +19560,8 @@ public static MyCollection<T> Create<T>(ReadOnlySpan<T> items)
                         => default;
                 }
                 """;
-            string sourceB = """
+
+            string sourceB1 = """
                 #pragma warning disable 219
                 class Program
                 {
@@ -19219,7 +19573,7 @@ static void Main()
                     }
                 }
                 """;
-            var comp = CreateCompilation(new[] { sourceA, sourceB }, targetFramework: TargetFramework.Net80);
+            var comp = CreateCompilation(new[] { sourceA, sourceB1 }, targetFramework: TargetFramework.Net80);
             comp.VerifyEmitDiagnostics(
                 // 1.cs(7,22): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
                 //         MyCollection<int> y = [1, 2, 3];
@@ -19227,6 +19581,45 @@ static void Main()
                 // 0.cs(15,35): error CS0452: The type 'T' must be a reference type in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
                 //     public static MyCollection<T> Create<T>(ReadOnlySpan<T> items)
                 Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "Create").WithArguments("MyCollection<T>", "T", "T").WithLocation(15, 35));
+
+            string sourceB2 = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        NoConstraints<int>([1]);
+                        NoConstraints<object>([2]);
+                        StructConstraint<int>([3]);
+                        ClassConstraint<object>([4]);
+                    }
+                    static MyCollection<T> NoConstraints<T>(params MyCollection<T> c) => c;
+                    static MyCollection<T> StructConstraint<T>(params MyCollection<T> c) where T : struct => c;
+                    static MyCollection<T> ClassConstraint<T>(params MyCollection<T> c) where T : class => c;
+                }
+                """;
+            comp = CreateCompilation(new[] { sourceA, sourceB2 }, targetFramework: TargetFramework.Net80);
+            comp.VerifyEmitDiagnostics(
+                // 1.cs(5,9): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
+                //         NoConstraints<int>([1]);
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "NoConstraints<int>").WithArguments("MyCollection<T>", "T", "int").WithLocation(5, 9),
+                // 1.cs(7,9): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
+                //         StructConstraint<int>([3]);
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "StructConstraint<int>").WithArguments("MyCollection<T>", "T", "int").WithLocation(7, 9),
+                // 1.cs(10,28): error CS0452: The type 'T' must be a reference type in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
+                //     static MyCollection<T> NoConstraints<T>(params MyCollection<T> c) => c;
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "NoConstraints").WithArguments("MyCollection<T>", "T", "T").WithLocation(10, 28),
+                // 1.cs(10,68): error CS0452: The type 'T' must be a reference type in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
+                //     static MyCollection<T> NoConstraints<T>(params MyCollection<T> c) => c;
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "c").WithArguments("MyCollection<T>", "T", "T").WithLocation(10, 68),
+                // 1.cs(11,28): error CS0452: The type 'T' must be a reference type in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
+                //     static MyCollection<T> StructConstraint<T>(params MyCollection<T> c) where T : struct => c;
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "StructConstraint").WithArguments("MyCollection<T>", "T", "T").WithLocation(11, 28),
+                // 1.cs(11,71): error CS0452: The type 'T' must be a reference type in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
+                //     static MyCollection<T> StructConstraint<T>(params MyCollection<T> c) where T : struct => c;
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "c").WithArguments("MyCollection<T>", "T", "T").WithLocation(11, 71),
+                // 0.cs(15,35): error CS0452: The type 'T' must be a reference type in order to use it as parameter 'T' in the generic type or method 'MyCollection<T>'
+                //     public static MyCollection<T> Create<T>(ReadOnlySpan<T> items)
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "Create").WithArguments("MyCollection<T>", "T", "T").WithLocation(15, 35));
         }
 
         [CombinatorialData]
@@ -19769,7 +20162,9 @@ static void Main()
                         MyCollection<int> x = [];
                         MyCollection<string> y = [null];
                         MyCollection<object> z = MyCollectionBuilder.Create<object>(default);
+                        x = Params(1, 2);
                     }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
                 }
                 """;
             var comp = CreateCompilation(sourceB, references: new[] { refA }, targetFramework: TargetFramework.Net80);
@@ -19782,7 +20177,10 @@ static void Main()
                 Diagnostic(ErrorCode.ERR_UnsupportedCompilerFeature, "[null]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "MyFeature").WithLocation(7, 34),
                 // (8,54): error CS9041: 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)' requires compiler feature 'MyFeature', which is not supported by this version of the C# compiler.
                 //         MyCollection<object> z = MyCollectionBuilder.Create<object>(default);
-                Diagnostic(ErrorCode.ERR_UnsupportedCompilerFeature, "Create<object>").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "MyFeature").WithLocation(8, 54));
+                Diagnostic(ErrorCode.ERR_UnsupportedCompilerFeature, "Create<object>").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "MyFeature").WithLocation(8, 54),
+                // (9,13): error CS9041: 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)' requires compiler feature 'MyFeature', which is not supported by this version of the C# compiler.
+                //         x = Params(1, 2);
+                Diagnostic(ErrorCode.ERR_UnsupportedCompilerFeature, "Params(1, 2)").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "MyFeature").WithLocation(9, 13));
         }
 
         [Fact]
@@ -19826,7 +20224,9 @@ static void Main()
                         MyCollection<int> x = [];
                         MyCollection<string> y = [null];
                         MyCollection<object> z = MyCollectionBuilder.Create<object>(default);
+                        x = Params(1, 2);
                     }
+                    static MyCollection<T> Params<T>(params MyCollection<T> c) => c;
                 }
                 """;
             var comp = CreateCompilation(sourceB, references: new[] { refA }, targetFramework: TargetFramework.Net80);
@@ -19842,7 +20242,10 @@ static void Main()
                 Diagnostic(ErrorCode.ERR_UnsupportedCompilerFeature, "MyCollectionBuilder").WithArguments("MyCollectionBuilder", "MyFeature").WithLocation(8, 34),
                 // (8,54): error CS9041: 'MyCollectionBuilder' requires compiler feature 'MyFeature', which is not supported by this version of the C# compiler.
                 //         MyCollection<object> z = MyCollectionBuilder.Create<object>(default);
-                Diagnostic(ErrorCode.ERR_UnsupportedCompilerFeature, "Create<object>").WithArguments("MyCollectionBuilder", "MyFeature").WithLocation(8, 54));
+                Diagnostic(ErrorCode.ERR_UnsupportedCompilerFeature, "Create<object>").WithArguments("MyCollectionBuilder", "MyFeature").WithLocation(8, 54),
+                // (9,13): error CS9041: 'MyCollectionBuilder' requires compiler feature 'MyFeature', which is not supported by this version of the C# compiler.
+                //         x = Params(1, 2);
+                Diagnostic(ErrorCode.ERR_UnsupportedCompilerFeature, "Params(1, 2)").WithArguments("MyCollectionBuilder", "MyFeature").WithLocation(9, 13));
         }
 
         [Fact]
@@ -21582,7 +21985,7 @@ .locals init (System.Collections.Generic.List<int> V_0,
                   }
                   IL_008b:  ldloc.0
                   IL_008c:  ret
-                                }
+                }
                 """);
         }
 
@@ -27201,7 +27604,7 @@ .. GetConfig(),
                 }
                 """;
 
-            var comp = CreateCompilation(source);
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular13);
             comp.VerifyEmitDiagnostics(
                 // (4,52): error CS9215: Collection expression type 'Dictionary<string, object>' must have an instance or extension method 'Add' that can be called with a single argument.
                 //     Dictionary<string, object> Config => /*<bind>*/[
@@ -31386,9 +31789,9 @@ public static MyCollection<T> Create<T>(ReadOnlySpan<T> items) where T : notnull
             // We're missing diagnostics for the `where T : notnull` constraint on the Create method
             // Tracked by https://github.com/dotnet/roslyn/issues/68786
             CreateCompilation(src, targetFramework: TargetFramework.Net80).VerifyEmitDiagnostics(
-                // (7,28): warning CS8714: The type 'string?' cannot be used as type parameter 'T' in the generic type or method 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
-                // MyCollection<string?> x1 = [null]; // 1
-                Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterNotNullConstraint, "[null]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "T", "string?").WithLocation(7, 28),
+                //// (7,28): warning CS8714: The type 'string?' cannot be used as type parameter 'T' in the generic type or method 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T>)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
+                //// MyCollection<string?> x1 = [null]; // 1
+                //Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterNotNullConstraint, "[null]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T>)", "T", "string?").WithLocation(7, 28),
                 // (8,28): warning CS8625: Cannot convert null literal to non-nullable reference type.
                 // MyCollection<string> x2 = [null]; // 2
                 Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(8, 28)
@@ -31431,10 +31834,12 @@ public static MyCollection<T> Create<T>(ReadOnlySpan<T?> items) where T : notnul
                 }
                 """;
 
+            // We're missing diagnostics for the `where T : notnull` constraint on the Create method
+            // Tracked by https://github.com/dotnet/roslyn/issues/68786
             CreateCompilation(src, targetFramework: TargetFramework.Net80).VerifyEmitDiagnostics(
-                // (7,28): warning CS8714: The type 'string?' cannot be used as type parameter 'T' in the generic type or method 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T?>)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
-                // MyCollection<string?> x1 = [null];
-                Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterNotNullConstraint, "[null]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T?>)", "T", "string?").WithLocation(7, 28),
+                //// (7,28): warning CS8714: The type 'string?' cannot be used as type parameter 'T' in the generic type or method 'MyCollectionBuilder.Create<T>(ReadOnlySpan<T?>)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
+                //// MyCollection<string?> x1 = [null];
+                //Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterNotNullConstraint, "[null]").WithArguments("MyCollectionBuilder.Create<T>(System.ReadOnlySpan<T?>)", "T", "string?").WithLocation(7, 28),
                 // (8,28): warning CS8625: Cannot convert null literal to non-nullable reference type.
                 // MyCollection<string> x2 = [null];
                 Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(8, 28)
@@ -38239,8 +38644,10 @@ public void Add(object obj) { }
         }
 
         [WorkItem("https://github.com/dotnet/roslyn/issues/72461")]
-        [Fact]
-        public void Add_ParamsArray_01()
+        [Theory]
+        [CombinatorialData]
+        public void Add_ParamsArray_01(
+            [CombinatorialValues(LanguageVersion.CSharp12, LanguageVersionFacts.CSharpNext)] LanguageVersion languageVersion)
         {
             string source = """
                 using System;
@@ -38269,7 +38676,10 @@ static void Main()
                 }
                 """;
 
-            var comp = CreateCompilation(source, options: TestOptions.ReleaseExe);
+            var comp = CreateCompilation(
+                source,
+                parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion),
+                options: TestOptions.ReleaseExe);
             comp.VerifyEmitDiagnostics();
             var verifier = CompileAndVerify(comp, expectedOutput: "(a, b), (c, d), ");
 
@@ -38318,8 +38728,10 @@ .locals init (T[] V_0,
         }
 
         [WorkItem("https://github.com/dotnet/roslyn/issues/72461")]
-        [Fact]
-        public void Add_ParamsCollection_01()
+        [Theory]
+        [CombinatorialData]
+        public void Add_ParamsCollection_01(
+            [CombinatorialValues(LanguageVersion.CSharp13, LanguageVersionFacts.CSharpNext)] LanguageVersion languageVersion)
         {
             string source = """
                 using System;
@@ -38348,7 +38760,10 @@ static void Main()
                 }
                 """;
 
-            var comp = CreateCompilation(source, options: TestOptions.ReleaseExe);
+            var comp = CreateCompilation(
+                source,
+                parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion),
+                options: TestOptions.ReleaseExe);
             CompileAndVerify(comp, expectedOutput: "(a, b), (c, d), ").VerifyDiagnostics();
 
             VerifyOperationTreeForTest<CollectionExpressionSyntax>(comp,
@@ -43078,12 +43493,12 @@ static void Main()
                 // (11,20): warning CS8600: Converting null literal or possible null value to non-nullable type.
                 //         object x = null;
                 Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "null").WithLocation(11, 20),
-                // (13,25): warning CS8601: Possible null reference assignment.
+                // (13,23): warning CS8601: Possible null reference assignment.
                 //         object[] z = [..y];
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "y").WithLocation(13, 25),
-                // (14,16): warning CS8601: Possible null reference assignment.
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "..y").WithLocation(13, 23),
+                // (14,14): warning CS8601: Possible null reference assignment.
                 //         z = [..y];
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "y").WithLocation(14, 16));
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "..y").WithLocation(14, 14));
         }
 
         [Fact]
@@ -43105,9 +43520,9 @@ static void Main()
                 """;
             var comp = CreateCompilation(source);
             comp.VerifyEmitDiagnostics(
-                // (8,37): warning CS8601: Possible null reference assignment.
+                // (8,35): warning CS8601: Possible null reference assignment.
                 //         IEnumerable<object> y1 = [..x1];
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "x1").WithLocation(8, 37));
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "..x1").WithLocation(8, 35));
         }
 
         [Fact]
@@ -43183,9 +43598,9 @@ static void Main()
                 // (14,14): warning CS8601: Possible null reference assignment.
                 //         z = [x, ..F(x)];
                 Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "x").WithLocation(14, 14),
-                // (14,19): warning CS8601: Possible null reference assignment.
+                // (14,17): warning CS8601: Possible null reference assignment.
                 //         z = [x, ..F(x)];
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "F(x)").WithLocation(14, 19),
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "..F(x)").WithLocation(14, 17),
                 // (16,26): warning CS8604: Possible null reference argument for parameter 'x' in 'IEnumerable<string> Program.F<string>(string x)'.
                 //         z = [..F<string>(x)];
                 Diagnostic(ErrorCode.WRN_NullReferenceArgument, "x").WithArguments("x", "IEnumerable<string> Program.F<string>(string x)").WithLocation(16, 26));
