diff --git a/src/Features/Core/Portable/SignatureHelp/CommonSignatureHelpUtilities.cs b/src/Features/Core/Portable/SignatureHelp/CommonSignatureHelpUtilities.cs
index 38fbff9ee67..43fb9fc2eab 100644
--- a/src/Features/Core/Portable/SignatureHelp/CommonSignatureHelpUtilities.cs
+++ b/src/Features/Core/Portable/SignatureHelp/CommonSignatureHelpUtilities.cs
@@ -100,6 +100,22 @@ internal static TextSpan GetSignatureHelpSpan<TArgumentList>(
         return TextSpan.FromBounds(start, nextToken.SpanStart);
     }
 
+    internal static async Task<TSyntax?> TryGetSyntaxAsync<TSyntax>(
+        Document document,
+        int position,
+        SignatureHelpTriggerReason triggerReason,
+        Func<SyntaxToken, bool> isTriggerToken,
+        Func<TSyntax, SyntaxToken, bool> isArgumentListToken,
+        CancellationToken cancellationToken)
+        where TSyntax : SyntaxNode
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+        return TryGetSyntax(
+            root, position, syntaxFacts, triggerReason, isTriggerToken, isArgumentListToken, cancellationToken, out var syntax) ? syntax : null;
+    }
+
     internal static bool TryGetSyntax<TSyntax>(
         SyntaxNode root,
         int position,
