diff --git a/src/Features/CSharp/Portable/Completion/KeywordRecommenders/WithKeywordRecommender.cs b/src/Features/CSharp/Portable/Completion/KeywordRecommenders/WithKeywordRecommender.cs
index 4289f46d415..d34e81c9e0d 100644
--- a/src/Features/CSharp/Portable/Completion/KeywordRecommenders/WithKeywordRecommender.cs
+++ b/src/Features/CSharp/Portable/Completion/KeywordRecommenders/WithKeywordRecommender.cs
@@ -4,11 +4,27 @@
 
 using System.Threading;
 using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
 
 namespace Microsoft.CodeAnalysis.CSharp.Completion.KeywordRecommenders;
 
 internal sealed class WithKeywordRecommender() : AbstractSyntacticSingleKeywordRecommender(SyntaxKind.WithKeyword)
 {
     protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
-        => !context.IsInNonUserCode && context.IsIsOrAsOrSwitchOrWithExpressionContext;
+    {
+        if (context.IsInNonUserCode)
+            return false;
+
+        if (context.IsIsOrAsOrSwitchOrWithExpressionContext)
+            return true;
+
+        var targetToken = context.TargetToken;
+        if (targetToken.Kind() == SyntaxKind.OpenBracketToken &&
+            targetToken.Parent is CollectionExpressionSyntax)
+        {
+            return true;
+        }
+
+        return false;
+    }
 }
