diff --git a/src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs b/src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs
index 487389d7fa9..65040844868 100644
--- a/src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs
+++ b/src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs
@@ -2514,6 +2514,10 @@ or ErrorCode.ERR_InitInExtension
                 or ErrorCode.ERR_ModifierOnUnnamedReceiverParameter
                 or ErrorCode.ERR_ExtensionTypeNameDisallowed
                 or ErrorCode.ERR_ExpressionTreeContainsNamedArgumentOutOfPosition
+                or ErrorCode.ERR_CollectionExpressionKeyValuePairNotSupported
+                or ErrorCode.ERR_CollectionArgumentsMustBeFirst
+                or ErrorCode.ERR_CollectionArgumentsNotSupportedForType
+                or ErrorCode.ERR_CollectionArgumentsDynamicBinding
                     => false,
             };
 #pragma warning restore CS8524 // The switch expression does not handle some values of its input type (it is not exhaustive) involving an unnamed enum value.
