diff --git a/src/Features/CSharp/Portable/SignatureHelp/ObjectCreationExpressionSignatureHelpProvider_NormalType.cs b/src/Features/CSharp/Portable/SignatureHelp/ObjectCreationExpressionSignatureHelpProvider_NormalType.cs
index 01e6979432a..4bf51b82f40 100644
--- a/src/Features/CSharp/Portable/SignatureHelp/ObjectCreationExpressionSignatureHelpProvider_NormalType.cs
+++ b/src/Features/CSharp/Portable/SignatureHelp/ObjectCreationExpressionSignatureHelpProvider_NormalType.cs
@@ -5,7 +5,6 @@
 using System.Collections.Generic;
 using System.Linq;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
-using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.DocumentationComments;
 using Microsoft.CodeAnalysis.LanguageService;
 using Microsoft.CodeAnalysis.Shared.Extensions;
@@ -15,14 +14,13 @@ namespace Microsoft.CodeAnalysis.CSharp.SignatureHelp;
 
 internal sealed partial class ObjectCreationExpressionSignatureHelpProvider
 {
-    private static SignatureHelpItem ConvertNormalTypeConstructor(
+    public static SignatureHelpItem ConvertNormalTypeConstructor(
         IMethodSymbol constructor,
-        BaseObjectCreationExpressionSyntax objectCreationExpression,
+        int position,
         SemanticModel semanticModel,
         IStructuralTypeDisplayService structuralTypeDisplayService,
         IDocumentationCommentFormattingService documentationCommentFormattingService)
     {
-        var position = objectCreationExpression.SpanStart;
         var item = CreateItem(
             constructor, semanticModel, position,
             structuralTypeDisplayService,
