diff --git a/src/Workspaces/CSharpTest/Formatting/FormattingTests.cs b/src/Workspaces/CSharpTest/Formatting/FormattingTests.cs
index bf3025d67eb..e16cfe0a12a 100644
--- a/src/Workspaces/CSharpTest/Formatting/FormattingTests.cs
+++ b/src/Workspaces/CSharpTest/Formatting/FormattingTests.cs
@@ -12563,4 +12563,76 @@ public  void    M   (   )
             """,
             parseOptions: CSharpParseOptions.Default.WithLanguageVersion(LanguageVersionExtensions.CSharpNext));
     }
+
+    [Fact]
+    public async Task FormatWithElement1()
+    {
+        await AssertFormatAsync(
+            expected: """
+                var v = [with()];
+                """,
+            code: """
+                var v = [ with ( ) ];
+                """);
+    }
+
+    [Fact]
+    public async Task FormatWithElement2()
+    {
+        await AssertFormatAsync(
+            expected: """
+                var v = [with(1, 2)];
+                """,
+            code: """
+                var v = [ with ( 1 , 2 ) ];
+                """);
+    }
+
+    [Fact]
+    public async Task FormatWithElement3()
+    {
+        await AssertFormatAsync(
+            expected: """
+                var v = [with(1, 2), 3];
+                """,
+            code: """
+                var v = [ with ( 1 , 2 ) , 3];
+                """);
+    }
+
+    [Fact]
+    public async Task FormatKeyValuePair1()
+    {
+        await AssertFormatAsync(
+            expected: """
+                var v = [k: v];
+                """,
+            code: """
+                var v = [ k :  v ];
+                """);
+    }
+
+    [Fact]
+    public async Task FormatKeyValuePair2()
+    {
+        await AssertFormatAsync(
+            expected: """
+                var v = [k: v, k: v];
+                """,
+            code: """
+                var v = [ k :  v , k :  v ];
+                """);
+    }
+
+    [Fact]
+    public async Task FormatKeyValuePair3()
+    {
+        await AssertFormatAsync(
+            expected: """
+                var v = [null: 1 + 1, (x.y): from x in y select z];
+                """,
+            code: """
+                var v = [ null :  1  +  1 , ( x . y ) :  from   x    in   y    select  z ];
+                """);
+    }
 }
