diff --git a/src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs b/src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs
index 731dcf784e0..e82685f6dbb 100644
--- a/src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs
@@ -136,6 +136,13 @@ internal static Conversion CreateCollectionExpressionConversion(
                 new CollectionExpressionUncommonData(collectionExpressionTypeKind, elementType, constructor, constructorUsedInExpandedForm, elementConversions));
         }
 
+        internal static Conversion CreateKeyValuePairConversion(Conversion keyConversion, Conversion valueConversion)
+        {
+            return new Conversion(
+                ConversionKind.KeyValuePair,
+                new NestedUncommonData([keyConversion, valueConversion]));
+        }
+
         private Conversion(
             ConversionKind kind,
             UncommonData? uncommonData = null)
@@ -560,6 +567,22 @@ internal CollectionExpressionTypeKind GetCollectionExpressionTypeKind(out TypeSy
             return CollectionExpressionTypeKind.None;
         }
 
+        internal bool TryGetKeyValueConversions(out Conversion keyConversion, out Conversion valueConversion)
+        {
+            if (Kind == ConversionKind.KeyValuePair)
+            {
+                Debug.Assert(_uncommonData is NestedUncommonData { _nestedConversionsOpt: [_, _] });
+                var nestedConversions = ((NestedUncommonData)_uncommonData)._nestedConversionsOpt;
+                keyConversion = nestedConversions[0];
+                valueConversion = nestedConversions[1];
+                return true;
+            }
+
+            keyConversion = NoConversion;
+            valueConversion = NoConversion;
+            return false;
+        }
+
         // CONSIDER: public?
         internal bool IsValid
         {
