diff --git a/src/Compilers/CSharp/Test/Syntax/Generated/Syntax.Test.xml.Generated.cs b/src/Compilers/CSharp/Test/Syntax/Generated/Syntax.Test.xml.Generated.cs
index ce449253b89..873922ca4b7 100644
--- a/src/Compilers/CSharp/Test/Syntax/Generated/Syntax.Test.xml.Generated.cs
+++ b/src/Compilers/CSharp/Test/Syntax/Generated/Syntax.Test.xml.Generated.cs
@@ -223,6 +223,12 @@ private static Syntax.InternalSyntax.ExpressionElementSyntax GenerateExpressionE
         private static Syntax.InternalSyntax.SpreadElementSyntax GenerateSpreadElement()
             => InternalSyntaxFactory.SpreadElement(InternalSyntaxFactory.Token(SyntaxKind.DotDotToken), GenerateIdentifierName());
 
+        private static Syntax.InternalSyntax.KeyValuePairElementSyntax GenerateKeyValuePairElement()
+            => InternalSyntaxFactory.KeyValuePairElement(GenerateIdentifierName(), InternalSyntaxFactory.Token(SyntaxKind.ColonToken), GenerateIdentifierName());
+
+        private static Syntax.InternalSyntax.WithElementSyntax GenerateWithElement()
+            => InternalSyntaxFactory.WithElement(InternalSyntaxFactory.Token(SyntaxKind.WithKeyword), GenerateArgumentList());
+
         private static Syntax.InternalSyntax.QueryExpressionSyntax GenerateQueryExpression()
             => InternalSyntaxFactory.QueryExpression(GenerateFromClause(), GenerateQueryBody());
 
@@ -1588,6 +1594,29 @@ public void TestSpreadElementFactoryAndProperties()
             AttachAndCheckDiagnostics(node);
         }
 
+        [Fact]
+        public void TestKeyValuePairElementFactoryAndProperties()
+        {
+            var node = GenerateKeyValuePairElement();
+
+            Assert.NotNull(node.KeyExpression);
+            Assert.Equal(SyntaxKind.ColonToken, node.ColonToken.Kind);
+            Assert.NotNull(node.ValueExpression);
+
+            AttachAndCheckDiagnostics(node);
+        }
+
+        [Fact]
+        public void TestWithElementFactoryAndProperties()
+        {
+            var node = GenerateWithElement();
+
+            Assert.Equal(SyntaxKind.WithKeyword, node.WithKeyword.Kind);
+            Assert.NotNull(node.ArgumentList);
+
+            AttachAndCheckDiagnostics(node);
+        }
+
         [Fact]
         public void TestQueryExpressionFactoryAndProperties()
         {
@@ -5762,6 +5791,58 @@ public void TestSpreadElementIdentityRewriter()
             Assert.Same(oldNode, newNode);
         }
 
+        [Fact]
+        public void TestKeyValuePairElementTokenDeleteRewriter()
+        {
+            var oldNode = GenerateKeyValuePairElement();
+            var rewriter = new TokenDeleteRewriter();
+            var newNode = rewriter.Visit(oldNode);
+
+            if(!oldNode.IsMissing)
+            {
+                Assert.NotEqual(oldNode, newNode);
+            }
+
+            Assert.NotNull(newNode);
+            Assert.True(newNode.IsMissing, "No tokens => missing");
+        }
+
+        [Fact]
+        public void TestKeyValuePairElementIdentityRewriter()
+        {
+            var oldNode = GenerateKeyValuePairElement();
+            var rewriter = new IdentityRewriter();
+            var newNode = rewriter.Visit(oldNode);
+
+            Assert.Same(oldNode, newNode);
+        }
+
+        [Fact]
+        public void TestWithElementTokenDeleteRewriter()
+        {
+            var oldNode = GenerateWithElement();
+            var rewriter = new TokenDeleteRewriter();
+            var newNode = rewriter.Visit(oldNode);
+
+            if(!oldNode.IsMissing)
+            {
+                Assert.NotEqual(oldNode, newNode);
+            }
+
+            Assert.NotNull(newNode);
+            Assert.True(newNode.IsMissing, "No tokens => missing");
+        }
+
+        [Fact]
+        public void TestWithElementIdentityRewriter()
+        {
+            var oldNode = GenerateWithElement();
+            var rewriter = new IdentityRewriter();
+            var newNode = rewriter.Visit(oldNode);
+
+            Assert.Same(oldNode, newNode);
+        }
+
         [Fact]
         public void TestQueryExpressionTokenDeleteRewriter()
         {
@@ -10530,6 +10611,12 @@ private static ExpressionElementSyntax GenerateExpressionElement()
         private static SpreadElementSyntax GenerateSpreadElement()
             => SyntaxFactory.SpreadElement(SyntaxFactory.Token(SyntaxKind.DotDotToken), GenerateIdentifierName());
 
+        private static KeyValuePairElementSyntax GenerateKeyValuePairElement()
+            => SyntaxFactory.KeyValuePairElement(GenerateIdentifierName(), SyntaxFactory.Token(SyntaxKind.ColonToken), GenerateIdentifierName());
+
+        private static WithElementSyntax GenerateWithElement()
+            => SyntaxFactory.WithElement(SyntaxFactory.Token(SyntaxKind.WithKeyword), GenerateArgumentList());
+
         private static QueryExpressionSyntax GenerateQueryExpression()
             => SyntaxFactory.QueryExpression(GenerateFromClause(), GenerateQueryBody());
 
@@ -11895,6 +11982,29 @@ public void TestSpreadElementFactoryAndProperties()
             Assert.Equal(node, newNode);
         }
 
+        [Fact]
+        public void TestKeyValuePairElementFactoryAndProperties()
+        {
+            var node = GenerateKeyValuePairElement();
+
+            Assert.NotNull(node.KeyExpression);
+            Assert.Equal(SyntaxKind.ColonToken, node.ColonToken.Kind());
+            Assert.NotNull(node.ValueExpression);
+            var newNode = node.WithKeyExpression(node.KeyExpression).WithColonToken(node.ColonToken).WithValueExpression(node.ValueExpression);
+            Assert.Equal(node, newNode);
+        }
+
+        [Fact]
+        public void TestWithElementFactoryAndProperties()
+        {
+            var node = GenerateWithElement();
+
+            Assert.Equal(SyntaxKind.WithKeyword, node.WithKeyword.Kind());
+            Assert.NotNull(node.ArgumentList);
+            var newNode = node.WithWithKeyword(node.WithKeyword).WithArgumentList(node.ArgumentList);
+            Assert.Equal(node, newNode);
+        }
+
         [Fact]
         public void TestQueryExpressionFactoryAndProperties()
         {
@@ -16069,6 +16179,58 @@ public void TestSpreadElementIdentityRewriter()
             Assert.Same(oldNode, newNode);
         }
 
+        [Fact]
+        public void TestKeyValuePairElementTokenDeleteRewriter()
+        {
+            var oldNode = GenerateKeyValuePairElement();
+            var rewriter = new TokenDeleteRewriter();
+            var newNode = rewriter.Visit(oldNode);
+
+            if(!oldNode.IsMissing)
+            {
+                Assert.NotEqual(oldNode, newNode);
+            }
+
+            Assert.NotNull(newNode);
+            Assert.True(newNode.IsMissing, "No tokens => missing");
+        }
+
+        [Fact]
+        public void TestKeyValuePairElementIdentityRewriter()
+        {
+            var oldNode = GenerateKeyValuePairElement();
+            var rewriter = new IdentityRewriter();
+            var newNode = rewriter.Visit(oldNode);
+
+            Assert.Same(oldNode, newNode);
+        }
+
+        [Fact]
+        public void TestWithElementTokenDeleteRewriter()
+        {
+            var oldNode = GenerateWithElement();
+            var rewriter = new TokenDeleteRewriter();
+            var newNode = rewriter.Visit(oldNode);
+
+            if(!oldNode.IsMissing)
+            {
+                Assert.NotEqual(oldNode, newNode);
+            }
+
+            Assert.NotNull(newNode);
+            Assert.True(newNode.IsMissing, "No tokens => missing");
+        }
+
+        [Fact]
+        public void TestWithElementIdentityRewriter()
+        {
+            var oldNode = GenerateWithElement();
+            var rewriter = new IdentityRewriter();
+            var newNode = rewriter.Visit(oldNode);
+
+            Assert.Same(oldNode, newNode);
+        }
+
         [Fact]
         public void TestQueryExpressionTokenDeleteRewriter()
         {
