diff --git a/src/Analyzers/Core/Analyzers/UseCollectionExpression/CollectionExpressionMatch.cs b/src/Analyzers/Core/Analyzers/UseCollectionExpression/CollectionExpressionMatch.cs
index db6671311ed..75560bd1e4c 100644
--- a/src/Analyzers/Core/Analyzers/UseCollectionExpression/CollectionExpressionMatch.cs
+++ b/src/Analyzers/Core/Analyzers/UseCollectionExpression/CollectionExpressionMatch.cs
@@ -16,4 +16,5 @@ namespace Microsoft.CodeAnalysis.UseCollectionExpression;
 /// what form it is.</param>
 internal readonly record struct CollectionMatch<TMatchNode>(
     TMatchNode Node,
-    bool UseSpread) where TMatchNode : SyntaxNode;
+    bool UseSpread,
+    bool UseKeyValue) where TMatchNode : SyntaxNode;
