diff --git a/src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs b/src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs
index 6ba23a9cfc2..f67472d9e32 100644
--- a/src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs
+++ b/src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs
@@ -931,5 +931,7 @@ public enum SyntaxKind : ushort
         ExtensionDeclaration = 9079,
 
         IgnoredDirectiveTrivia = 9080,
+        KeyValuePairElement = 9081,
+        WithElement = 9082,
     }
 }
