diff --git a/src/Compilers/CSharp/Test/Emit3/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs b/src/Compilers/CSharp/Test/Emit3/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs
index 77f2542f3d2..3b87c112b52 100644
--- a/src/Compilers/CSharp/Test/Emit3/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs
+++ b/src/Compilers/CSharp/Test/Emit3/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs
@@ -33,9 +33,6 @@ public void DiagnosticAnalyzerAllInOne()
             // https://github.com/dotnet/roslyn/issues/44682 Add to all in one
             missingSyntaxKinds.Add(SyntaxKind.WithExpression);
             missingSyntaxKinds.Add(SyntaxKind.RecordDeclaration);
-            missingSyntaxKinds.Add(SyntaxKind.CollectionExpression);
-            missingSyntaxKinds.Add(SyntaxKind.ExpressionElement);
-            missingSyntaxKinds.Add(SyntaxKind.SpreadElement);
             // Tracked by https://github.com/dotnet/roslyn/issues/76130 : Add to all-in-one
             missingSyntaxKinds.Add(SyntaxKind.ExtensionDeclaration);
 
