diff --git a/src/EditorFeatures/TestUtilities/SignatureHelp/AbstractSignatureHelpProviderTests.cs b/src/EditorFeatures/TestUtilities/SignatureHelp/AbstractSignatureHelpProviderTests.cs
index 5bf6810b8e0..d6f1ef03248 100644
--- a/src/EditorFeatures/TestUtilities/SignatureHelp/AbstractSignatureHelpProviderTests.cs
+++ b/src/EditorFeatures/TestUtilities/SignatureHelp/AbstractSignatureHelpProviderTests.cs
@@ -72,21 +72,14 @@ private async Task TestSignatureHelpWorkerAsync(
         bool usePreviousCharAsTrigger = false)
     {
         using var workspaceFixture = GetOrCreateWorkspaceFixture();
+
+        var workspace = workspaceFixture.Target.GetWorkspace(markupWithPositionAndOptSpan);
+
         var options = new MemberDisplayOptions();
 
-        markupWithPositionAndOptSpan = markupWithPositionAndOptSpan.NormalizeLineEndings();
-
-        TextSpan? textSpan = null;
-        MarkupTestFile.GetPositionAndSpans(
-            markupWithPositionAndOptSpan,
-            out var code,
-            out var cursorPosition,
-            out var textSpans);
-
-        if (textSpans.Any())
-        {
-            textSpan = textSpans.First();
-        }
+        var code = workspaceFixture.Target.Code;
+        var position = workspaceFixture.Target.Position;
+        var textSpan = workspaceFixture.Target.Spans.FirstOrNull();
 
         var parseOptions = CreateExperimentalParseOptions();
 
@@ -97,10 +90,10 @@ private async Task TestSignatureHelpWorkerAsync(
             document1 = document1.Project.WithParseOptions(parseOptions).GetDocument(document1.Id);
         }
 
-        await TestSignatureHelpWorkerSharedAsync(workspaceFixture.Target.GetWorkspace(), code, cursorPosition, document1, options, textSpan, expectedOrderedItemsOrNull, usePreviousCharAsTrigger);
+        await TestSignatureHelpWorkerSharedAsync(workspaceFixture.Target.GetWorkspace(), code, position, document1, options, textSpan, expectedOrderedItemsOrNull, usePreviousCharAsTrigger);
 
         // speculative semantic model
-        if (await CanUseSpeculativeSemanticModelAsync(document1, cursorPosition))
+        if (await CanUseSpeculativeSemanticModelAsync(document1, position))
         {
             var document2 = workspaceFixture.Target.UpdateDocument(code, sourceCodeKind, cleanBeforeUpdate: false);
             if (experimental)
@@ -108,7 +101,7 @@ private async Task TestSignatureHelpWorkerAsync(
                 document2 = document2.Project.WithParseOptions(parseOptions).GetDocument(document2.Id);
             }
 
-            await TestSignatureHelpWorkerSharedAsync(workspaceFixture.Target.GetWorkspace(), code, cursorPosition, document2, options, textSpan, expectedOrderedItemsOrNull, usePreviousCharAsTrigger);
+            await TestSignatureHelpWorkerSharedAsync(workspaceFixture.Target.GetWorkspace(), code, position, document2, options, textSpan, expectedOrderedItemsOrNull, usePreviousCharAsTrigger);
         }
     }
 
