diff --git a/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs b/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs
index 6739e439ee5..1e7c0c78c29 100644
--- a/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs
@@ -684,13 +684,21 @@ private void MakeCollectionExpressionTypeInferences(
 
             foreach (var element in argument.Elements)
             {
-                if (element is BoundCollectionExpressionSpreadElement spread)
+                switch (element)
                 {
-                    MakeSpreadElementTypeInferences(spread, targetElementType, ref useSiteInfo);
-                }
-                else
-                {
-                    MakeExplicitParameterTypeInferences(binder, (BoundExpression)element, targetElementType, kind, ref useSiteInfo);
+                    case BoundCollectionExpressionWithElement withElement:
+                        // Arguments do not affect type inference.
+                        break;
+                    case BoundCollectionExpressionSpreadElement spread:
+                        MakeSpreadElementTypeInferences(spread, targetElementType, ref useSiteInfo);
+                        break;
+                    case BoundKeyValuePairElement keyValuePairElement:
+                    case BoundKeyValuePairConversion keyValuePairConversion:
+                        // https://github.com/dotnet/roslyn/issues/77873: Handle input type inference for key:value elements.
+                        break;
+                    default:
+                        MakeExplicitParameterTypeInferences(binder, (BoundExpression)element, targetElementType, kind, ref useSiteInfo);
+                        break;
                 }
             }
         }
@@ -909,6 +917,7 @@ private void MakeOutputTypeInferences(Binder binder, BoundUnconvertedCollectionE
 
             foreach (var element in argument.Elements)
             {
+                // https://github.com/dotnet/roslyn/issues/77873: Handle output type inference for key:value elements.
                 if (element is BoundExpression expression)
                 {
                     MakeOutputTypeInferences(binder, expression, targetElementType, ref useSiteInfo);
