diff --git a/src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpUseCollectionExpressionForFluentCodeFixProvider.cs b/src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpUseCollectionExpressionForFluentCodeFixProvider.cs
index 0d6ebea4aa3..6436b15214f 100644
--- a/src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpUseCollectionExpressionForFluentCodeFixProvider.cs
+++ b/src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpUseCollectionExpressionForFluentCodeFixProvider.cs
@@ -25,8 +25,8 @@
 namespace Microsoft.CodeAnalysis.CSharp.UseCollectionExpression;
 
 using static CSharpCollectionExpressionRewriter;
-using static CSharpUseCollectionExpressionForFluentDiagnosticAnalyzer;
 using static CSharpSyntaxTokens;
+using static CSharpUseCollectionExpressionForFluentDiagnosticAnalyzer;
 using static SyntaxFactory;
 
 [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.UseCollectionExpressionForFluent), Shared]
@@ -59,10 +59,10 @@ protected override async Task FixAsync(
 
         // We want to replace `new[] { 1, 2, 3 }.Concat(x).Add(y).ToArray()` with the new collection expression.  To do
         // this, we go through the following steps.  First, we replace the whole expression with `new(x, y) { 1, 2, 3 }`
-        // (a dummy object creation expression). We then call into our helper which replaces expressions with
-        // collection expressions.  The reason for the dummy object creation expression is that it serves as an actual
-        // node the rewriting code can attach an initializer to, by which it can figure out appropriate wrapping and
-        // indentation for the collection expression elements.
+        // (a dummy object creation expression). We then call into our helper which replaces expressions with collection
+        // expressions.  The reason for the dummy object creation expression is that it serves as an actual node the
+        // rewriting code can attach an initializer to, by which it can figure out appropriate wrapping and indentation
+        // for the collection expression elements.
 
         var semanticDocument = await SemanticDocument.CreateAsync(document, cancellationToken).ConfigureAwait(false);
 
@@ -119,17 +119,17 @@ static ImmutableArray<CollectionMatch<ExpressionSyntax>> CreateMatches(
                     {
                         if (element is SpreadElementSyntax spreadElement)
                         {
-                            result.Add(new(spreadElement.Expression, UseSpread: true));
+                            result.Add(new(spreadElement.Expression, UseSpread: true, UseKeyValue: false));
                         }
                         else if (element is ExpressionElementSyntax expressionElement)
                         {
-                            result.Add(new(expressionElement.Expression, UseSpread: false));
+                            result.Add(new(expressionElement.Expression, UseSpread: false, UseKeyValue: false));
                         }
                     }
                 }
                 else
                 {
-                    result.Add(new(argument.Expression, match.UseSpread));
+                    result.Add(new(argument.Expression, match.UseSpread, UseKeyValue: false));
                 }
             }
 
