diff --git a/src/Features/Core/Portable/SignatureHelp/AbstractSignatureHelpProvider.cs b/src/Features/Core/Portable/SignatureHelp/AbstractSignatureHelpProvider.cs
index d2b348584b2..0fe11f043bd 100644
--- a/src/Features/Core/Portable/SignatureHelp/AbstractSignatureHelpProvider.cs
+++ b/src/Features/Core/Portable/SignatureHelp/AbstractSignatureHelpProvider.cs
@@ -175,7 +175,8 @@ protected static SignatureHelpItem CreateItemImpl(
         IList<SymbolDisplayPart> separatorParts,
         IList<SymbolDisplayPart> suffixParts,
         IList<SignatureHelpSymbolParameter> parameters,
-        IList<SymbolDisplayPart>? descriptionParts)
+        IList<SymbolDisplayPart>? descriptionParts,
+        Func<ISymbol?, string?>? getNavigationHint = null)
     {
         descriptionParts = descriptionParts == null
             ? SpecializedCollections.EmptyList<SymbolDisplayPart>()
@@ -218,11 +219,11 @@ where part.Symbol.IsAnonymousType() || part.Symbol.IsTupleType()
             orderSymbol,
             isVariadic,
             documentationFactory,
-            prefixParts.ToTaggedText(),
-            separatorParts.ToTaggedText(),
-            suffixParts.ToTaggedText(),
+            prefixParts.ToTaggedText(getNavigationHint: getNavigationHint),
+            separatorParts.ToTaggedText(getNavigationHint: getNavigationHint),
+            suffixParts.ToTaggedText(getNavigationHint: getNavigationHint),
             parameters.Select(p => (SignatureHelpParameter)p),
-            descriptionParts.ToTaggedText());
+            descriptionParts.ToTaggedText(getNavigationHint: getNavigationHint));
     }
 
     private static SignatureHelpSymbolParameter ReplaceStructuralTypes(
