diff --git a/src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs b/src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs
index 3ad3e5986aa..2a1ffce4b12 100644
--- a/src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs
+++ b/src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs
@@ -10,6 +10,7 @@
 using System.Linq;
 using System.Threading;
 using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Shared.Extensions;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.CSharp.Utilities;
 using Microsoft.CodeAnalysis.Operations;
@@ -90,6 +91,9 @@ public static bool CanReplaceWithCollectionExpression(
         var compilation = semanticModel.Compilation;
         changesSemantics = false;
 
+        if (!semanticModel.Compilation.LanguageVersion().SupportsCollectionExpressions())
+            return false;
+
         var topMostExpression = expression.WalkUpParentheses();
         if (topMostExpression.GetDiagnostics().Any(d => d.Severity == DiagnosticSeverity.Error))
             return false;
@@ -101,7 +105,7 @@ public static bool CanReplaceWithCollectionExpression(
             return false;
 
         // (X[])[1, 2, 3] is target typed.  `(X)[1, 2, 3]` is currently not (because it looks like indexing into an expr).
-        if (topMostExpression.Parent is CastExpressionSyntax castExpression && castExpression.Type is IdentifierNameSyntax)
+        if (topMostExpression.Parent is CastExpressionSyntax { Type: IdentifierNameSyntax })
             return false;
 
         // X[] = new Y[] { 1, 2, 3 }
@@ -116,7 +120,7 @@ public static bool CanReplaceWithCollectionExpression(
         if (originalTypeInfo.ConvertedType is null or IErrorTypeSymbol)
             return false;
 
-        if (!IsConstructibleCollectionType(compilation, originalTypeInfo.ConvertedType.OriginalDefinition))
+        if (!CollectionExpressionUtilities.IsConstructibleCollectionType(compilation, originalTypeInfo.ConvertedType.OriginalDefinition))
             return false;
 
         if (expression.IsInExpressionTree(semanticModel, expressionType, cancellationToken))
@@ -143,7 +147,7 @@ public static bool CanReplaceWithCollectionExpression(
         // expression will always succeed, and there's no need to actually validate semantics there.
         // Tracked by https://github.com/dotnet/roslyn/issues/68826
         if (parent is CastExpressionSyntax)
-            return IsConstructibleCollectionType(compilation, semanticModel.GetTypeInfo(parent, cancellationToken).Type);
+            return CollectionExpressionUtilities.IsConstructibleCollectionType(compilation, semanticModel.GetTypeInfo(parent, cancellationToken).Type);
 
         // Looks good as something to replace.  Now check the semantics of making the replacement to see if there would
         // any issues.
@@ -230,14 +234,23 @@ bool IsSafeConversionWhenTypesDoNotMatch(out bool changesSemantics)
             if (s_tupleNamesCanDifferComparer.Equals(type, convertedType))
                 return true;
 
-            // It's always safe to convert List<X> to ICollection<X> or IList<X> as the language guarantees that it will
-            // continue emitting a List<X> for those target types.
-            var isWellKnownCollectionReadWriteInterface = IsWellKnownCollectionReadWriteInterface(convertedType);
-            if (isWellKnownCollectionReadWriteInterface &&
-                Equals(type.OriginalDefinition, compilation.ListOfTType()) &&
-                type.AllInterfaces.Contains(convertedType))
+            // It is always safe to convert List<X> to ICollection<X> or IList<X> as the language guarantees that it
+            // will continue emitting a List<X> for those target types.
+            //
+            // Similarly, it is safe to convert Dictionary<X,Y> to IDictionary<X,Y> in C# 14 and above as the language
+            // guarantees that it will continue emitting a Dictionary<X,Y> for those target types.
+            var isWellKnownCollectionReadWriteInterface = CollectionExpressionUtilities.IsWellKnownCollectionReadWriteInterface(
+                compilation, convertedType);
+            if (isWellKnownCollectionReadWriteInterface)
             {
-                return true;
+                if (type.AllInterfaces.Contains(convertedType))
+                {
+                    if (Equals(type.OriginalDefinition, compilation.ListOfTType()))
+                        return true;
+
+                    if (Equals(type.OriginalDefinition, compilation.DictionaryOfKVType()))
+                        return compilation.LanguageVersion().SupportsDictionaryExpressions();
+                }
             }
 
             // Before this point are all the changes that we can detect that are always safe to make.
@@ -258,7 +271,8 @@ bool IsSafeConversionWhenTypesDoNotMatch(out bool changesSemantics)
             //
             // `IEnumerable<object> obj = Array.Empty<object>();` or
             // `IEnumerable<string> obj = new[] { "" };`
-            if (IsWellKnownCollectionInterface(convertedType) && type.AllInterfaces.Contains(convertedType))
+            if (CollectionExpressionUtilities.IsWellKnownCollectionInterface(compilation, convertedType) &&
+                type.AllInterfaces.Contains(convertedType))
             {
                 // The observable collections are known to have significantly different behavior than List<T>.  So
                 // disallow converting those types to ensure semantics are preserved.  We do this even though
@@ -296,91 +310,6 @@ bool IsSafeConversionWhenTypesDoNotMatch(out bool changesSemantics)
         }
     }
 
-    public static bool IsWellKnownCollectionInterface(ITypeSymbol type)
-        => IsWellKnownCollectionReadOnlyInterface(type) || IsWellKnownCollectionReadWriteInterface(type);
-
-    public static bool IsWellKnownCollectionReadOnlyInterface(ITypeSymbol type)
-    {
-        return type.OriginalDefinition.SpecialType
-            is SpecialType.System_Collections_Generic_IEnumerable_T
-            or SpecialType.System_Collections_Generic_IReadOnlyCollection_T
-            or SpecialType.System_Collections_Generic_IReadOnlyList_T;
-    }
-
-    public static bool IsWellKnownCollectionReadWriteInterface(ITypeSymbol type)
-    {
-        return type.OriginalDefinition.SpecialType
-            is SpecialType.System_Collections_Generic_ICollection_T
-            or SpecialType.System_Collections_Generic_IList_T;
-    }
-
-    public static bool IsConstructibleCollectionType(Compilation compilation, ITypeSymbol? type)
-    {
-        if (type is null)
-            return false;
-
-        // Arrays are always a valid collection expression type.
-        if (type is IArrayTypeSymbol)
-            return true;
-
-        // Has to be a real named type at this point.
-        if (type is INamedTypeSymbol namedType)
-        {
-            // Span<T> and ReadOnlySpan<T> are always valid collection expression types.
-            if (namedType.OriginalDefinition.Equals(compilation.SpanOfTType()) ||
-                namedType.OriginalDefinition.Equals(compilation.ReadOnlySpanOfTType()))
-            {
-                return true;
-            }
-
-            // If it has a [CollectionBuilder] attribute on it, it is a valid collection expression type.
-            if (namedType.GetAttributes().Any(a => a.AttributeClass.IsCollectionBuilderAttribute()))
-                return true;
-
-            if (IsWellKnownCollectionInterface(namedType))
-                return true;
-
-            // At this point, all that is left are collection-initializer types.  These need to derive from
-            // System.Collections.IEnumerable, and have an invokable no-arg constructor.
-
-            // Abstract type don't have invokable constructors at all.
-            if (namedType.IsAbstract)
-                return false;
-
-            if (namedType.AllInterfaces.Contains(compilation.IEnumerableType()!))
-            {
-                // If they have an accessible `public C(int capacity)` constructor, the lang prefers calling that.
-                var constructors = namedType.Constructors;
-                var capacityConstructor = GetAccessibleInstanceConstructor(constructors, c => c.Parameters is [{ Name: "capacity", Type.SpecialType: SpecialType.System_Int32 }]);
-                if (capacityConstructor != null)
-                    return true;
-
-                var noArgConstructor =
-                    GetAccessibleInstanceConstructor(constructors, c => c.Parameters.IsEmpty) ??
-                    GetAccessibleInstanceConstructor(constructors, c => c.Parameters.All(p => p.IsOptional || p.IsParams));
-                if (noArgConstructor != null)
-                {
-                    // If we have a struct, and the constructor we find is implicitly declared, don't consider this
-                    // a constructible type.  It's likely the user would just get the `default` instance of the
-                    // collection (like with ImmutableArray<T>) which would then not actually work.  If the struct
-                    // does have an explicit constructor though, that's a good sign it can actually be constructed
-                    // safely with the no-arg `new S()` call.
-                    if (!(namedType.TypeKind == TypeKind.Struct && noArgConstructor.IsImplicitlyDeclared))
-                        return true;
-                }
-            }
-        }
-
-        // Anything else is not constructible.
-        return false;
-
-        IMethodSymbol? GetAccessibleInstanceConstructor(ImmutableArray<IMethodSymbol> constructors, Func<IMethodSymbol, bool> predicate)
-        {
-            var constructor = constructors.FirstOrDefault(c => !c.IsStatic && predicate(c));
-            return constructor is not null && constructor.IsAccessibleWithin(compilation.Assembly) ? constructor : null;
-        }
-    }
-
     private static bool IsSafeConversionOfArrayToSpanType(
         SemanticModel semanticModel, ExpressionSyntax expression, CancellationToken cancellationToken)
     {
@@ -627,12 +556,12 @@ bool IsPrimitiveConstant(ExpressionSyntax expression)
     }
 
     public static CollectionExpressionSyntax ConvertInitializerToCollectionExpression(
-        InitializerExpressionSyntax initializer, bool wasOnSingleLine)
+        SourceText text, InitializerExpressionSyntax initializer, bool wasOnSingleLine)
     {
         // if the initializer is already on multiple lines, keep it that way.  otherwise, squash from `{ 1, 2, 3 }` to `[1, 2, 3]`
         var openBracket = OpenBracketToken.WithTriviaFrom(initializer.OpenBraceToken);
         var elements = initializer.Expressions.GetWithSeparators().SelectAsArray(
-            i => i.IsToken ? i : ExpressionElement((ExpressionSyntax)i.AsNode()!));
+            i => i.IsToken ? i : CreateElement((ExpressionSyntax)i.AsNode()!));
         var closeBracket = CloseBracketToken.WithTriviaFrom(initializer.CloseBraceToken);
 
         // If it was on a single line to begin with, then remove the inner spaces on the `{ ... }` to create `[...]`. If
@@ -648,6 +577,43 @@ public static CollectionExpressionSyntax ConvertInitializerToCollectionExpressio
         }
 
         return CollectionExpression(openBracket, SeparatedList<CollectionElementSyntax>(elements), closeBracket);
+
+        CollectionElementSyntax CreateElement(ExpressionSyntax expression)
+        {
+            if (expression is InitializerExpressionSyntax { Expressions: [var keyExpression1, var valueExpression1] } initializer)
+            {
+                // If we have `{ key, ... }` we want to move the leading trivia of the `{` to the key so that it is
+                // properly indented to the same level the `{` was.
+                var openBraceAndKeyOnSingleLine = wasOnSingleLine || text.AreOnSameLine(initializer.OpenBraceToken, keyExpression1.GetLastToken());
+                if (openBraceAndKeyOnSingleLine)
+                    keyExpression1 = keyExpression1.WithLeadingTrivia(initializer.OpenBraceToken.LeadingTrivia);
+
+                // If we have `{ ..., value }` we want to move the trailing trivia of the `}` to the value to preserve trailing comments.
+                var valueAndCloseBraceOnSingleLine = wasOnSingleLine || text.AreOnSameLine(valueExpression1.GetLastToken(), initializer.CloseBraceToken);
+                if (valueAndCloseBraceOnSingleLine)
+                    valueExpression1 = valueExpression1.WithTrailingTrivia(initializer.CloseBraceToken.TrailingTrivia);
+
+                return KeyValuePairElement(keyExpression1, ColonToken.WithTriviaFrom(initializer.Expressions.GetSeparator(0)), valueExpression1);
+            }
+            else if (expression is AssignmentExpressionSyntax
+            {
+                Left: ImplicitElementAccessSyntax { ArgumentList.Arguments: [var argument] } implicitElementAccess,
+                OperatorToken: var equalsToken,
+                Right: var valueExpression2,
+            })
+            {
+                // If we have `[key] = value` we want to move the leading trivia of the `[` to the key.
+                var keyExpression2 = argument.Expression.WithLeadingTrivia(implicitElementAccess.GetLeadingTrivia());
+                return KeyValuePairElement(
+                    keyExpression2,
+                    ColonToken.WithTrailingTrivia(equalsToken.TrailingTrivia),
+                    valueExpression2);
+            }
+            else
+            {
+                return ExpressionElement(expression);
+            }
+        }
     }
 
     public static CollectionExpressionSyntax ReplaceWithCollectionExpression(
@@ -836,7 +802,7 @@ public static ImmutableArray<CollectionMatch<StatementSyntax>> TryGetMatches<TAr
 
                         // this looks like a good statement, add to the right size of the assignment to track as that's what
                         // we'll want to put in the final collection expression.
-                        matches.Add(new(expressionStatement, UseSpread: false));
+                        matches.Add(new(expressionStatement, UseSpread: false, UseKeyValue: false));
                         currentStatement = currentStatement.GetNextStatement();
                     }
                 }
